<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Julien Bryois" />

<meta name="date" content="2022-01-28" />

<title>Plot figures</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">snRNA_eqtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Plot figures</h1>
<h4 class="author">Julien Bryois</h4>
<h4 class="date">January 28, 2022</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-01-28
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>snRNA_eqtl/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20211124code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20211124)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20211124code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20211124)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongf63c42b"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> f63c42b </a>
</p>
</div>
<div id="strongRepositoryversionstrongf63c42b" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version f63c42b. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/magma_interaction_eqtl/
    Ignored:    data/magma_specific_genes/
    Ignored:    data_sensitive/

Untracked files:
    Untracked:  analysis/._index.Rmd
    Untracked:  analysis/additional/
    Untracked:  analysis/alternatives/
    Untracked:  analysis/clean_analysis.Rmd
    Untracked:  analysis/log_run_slurm_out
    Untracked:  analysis/log_run_slurm_stderr
    Untracked:  analysis/portfolio/
    Untracked:  analysis/revision.Rmd
    Untracked:  analysis/run_slurm.Rscript
    Untracked:  analysis/run_slurm.sbatch
    Untracked:  data/dice/
    Untracked:  data/epigenome_enrichment/
    Untracked:  data/gencode/
    Untracked:  data/gnomad_loeuf/
    Untracked:  data/gtex/
    Untracked:  data/gwas/
    Untracked:  data/gwas_epigenome_overlap/
    Untracked:  data/metabrain/
    Untracked:  data/umap/
    Untracked:  output/coloc/
    Untracked:  output/eqtl/
    Untracked:  output/eqtl_specific/
    Untracked:  output/figures/
    Untracked:  output/figures_rev/
    Untracked:  output/gwas_epigenome_overlap/
    Untracked:  output/shiny/
    Untracked:  output/tables/
    Untracked:  output_tmp_QTLtools/
    Untracked:  output_tmp_fastQTL_sample_ambient_removed/

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/_site.yml
    Deleted:    output/README.md

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/plot_figures.Rmd</code>) and HTML (<code>docs/plot_figures.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
f63c42b
</td>
<td>
Julien Bryois
</td>
<td>
2022-01-28
</td>
<td>
plot figures
</td>
</tr>
<tr>
<td>
html
</td>
<td>
d151c10
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
254b41e
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
added_marker_figures - added code folding
</td>
</tr>
<tr>
<td>
html
</td>
<td>
19a5db5
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
5719848
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
added_marker_figures - small modif2
</td>
</tr>
<tr>
<td>
html
</td>
<td>
056f5c7
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
3d23918
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
added_marker_figures - small modif
</td>
</tr>
<tr>
<td>
html
</td>
<td>
811141f
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
a14eee2
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
added_marker_figures
</td>
</tr>
<tr>
<td>
html
</td>
<td>
d2896af
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
b1d314e
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
<td>
eqtl_code_added, figures added
</td>
</tr>
<tr>
<td>
html
</td>
<td>
cefc5e2
</td>
<td>
Julien Bryois
</td>
<td>
2021-11-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
45d6051
</td>
<td>
Julien Bryois
</td>
<td>
2021-11-26
</td>
<td>
Genotype process clean
</td>
</tr>
<tr>
<td>
html
</td>
<td>
5a68f72
</td>
<td>
Julien Bryois
</td>
<td>
2021-11-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
0ab5840
</td>
<td>
Julien Bryois
</td>
<td>
2021-11-26
</td>
<td>
Publish genotype processing + initial files
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(tidyverse)
library(qvalue)
library(patchwork)
library(scales)
library(liftOver)
library(ComplexHeatmap)
library(rtracklayer)
library(viridis)
library(rhdf5)</code></pre>
<div id="main-figures" class="section level1">
<h1>Main figures</h1>
<pre class="r"><code>dir.create(&#39;output/figures&#39;,showWarnings = FALSE)</code></pre>
<div id="figure-1" class="section level2">
<h2>Figure 1</h2>
<div id="umap" class="section level3">
<h3>UMAP</h3>
<pre class="r"><code>umap &lt;- read_tsv(&#39;data/umap/umap.ad.subsample.txt&#39;)</code></pre>
<pre class="r"><code>pos &lt;- umap %&gt;% group_by(broad) %&gt;% 
  summarise(UMAP_1=median(UMAP_1),
            UMAP_2=median(UMAP_2))</code></pre>
<pre class="r"><code>p &lt;- ggplot(umap,aes(UMAP_1,UMAP_2,col=broad)) + 
  geom_point(size=0.1) + 
  theme_classic() +
  theme(legend.position=&#39;none&#39;,
        axis.text = element_blank(),
        axis.ticks = element_blank(), 
        axis.title = element_blank(),
        axis.line = element_blank()) 
p + geom_text(data=pos,aes(label=broad),col=&#39;black&#39;,size=5)</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-6-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>p &lt;- p + geom_text(data=pos,aes(label=broad),col=&#39;black&#39;,size=9)
ggsave(p,filename = &#39;output/figures/Figure1_umap.png&#39;,width=12,height=8,dpi = 300)</code></pre>
</div>
<div id="interaction" class="section level3">
<h3>Interaction</h3>
<pre class="r"><code>hdF5_file_path &lt;- &#39;output/shiny/data.h5&#39;</code></pre>
<pre class="r"><code>eqtl_specific &lt;- h5read(hdF5_file_path, &quot;eqtl_results/eqtl_results_specific&quot;)</code></pre>
<pre class="r"><code>plot_eqtl_specific &lt;- function(cell_type_name,gene_name,snp_name,cells_to_display){
  
  gene_short &lt;- gsub(&#39;_.+&#39;,&#39;&#39;,gene_name)
  
  genotype &lt;- h5read(hdF5_file_path, paste0(&quot;genotype/&quot;,snp_name))
  expression &lt;- h5read(hdF5_file_path, paste0(&quot;expression/&quot;,gene_name)) %&gt;% 
    dplyr::rename(individual=individual_id) %&gt;% 
    mutate(cell_type=fct_relevel(cell_type,cell_type_name)) %&gt;% 
    filter(cell_type%in%cells_to_display)
  
  pval &lt;- filter(eqtl_specific,gene==gene_name,cell_type==cell_type_name) %&gt;% pull(nb_pvalue_all_adj)
  
  d &lt;- inner_join(genotype,expression,by=&#39;individual&#39;) %&gt;% 
  mutate(genotype_label = case_when(
    genotype == 0 ~ paste0(REF,REF),
    genotype == 1 ~ paste0(REF,ALT),
    genotype == 2 ~ paste0(ALT,ALT)
  )) 
  
  p &lt;- ggplot(d, aes(genotype_label,log2_cpm,col=cell_type)) + 
    ggbeeswarm::geom_quasirandom(size=0.5) + 
      geom_boxplot(alpha=0.05,aes(group=genotype_label),outlier.shape = NA) + 
    theme_classic() + 
    theme(legend.position=&#39;none&#39;,text=element_blank(),strip.background = element_blank(),
  strip.text.x = element_blank(),axis.text = element_blank(),axis.ticks = element_blank(), axis.line = element_blank()) + 
    xlab(&#39;&#39;) + 
    ylab(&#39;&#39;) +
 facet_wrap(~cell_type,ncol=4) +
    scale_color_discrete(limits=levels(d$cell_type) %&gt;% as.character() %&gt;% sort(),drop=FALSE)
  p
}</code></pre>
<pre class="r"><code>gene_name &lt;- &#39;CHRM5_ENSG00000184984&#39;
snp_name &lt;- &#39;rs8038713&#39;
cell_type_name &lt;- &#39;Oligodendrocytes&#39;
cells_to_display &lt;- c(&#39;Oligodendrocytes&#39;,&#39;Excitatory neurons&#39;, &#39;Inhibitory neurons&#39;)
chrm5 &lt;- plot_eqtl_specific(cell_type_name,gene_name,snp_name,cells_to_display) + theme(text=element_text(size=20),plot.title = element_text(size=12))
chrm5</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-10-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(chrm5,filename = &#39;output/figures/Figure1_int.png&#39;,width=7,height=3)</code></pre>
</div>
</div>
<div id="figure-2a" class="section level2">
<h2>Figure 2A</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;) %&gt;% 
  filter(adj_p&lt;0.05)

p &lt;- d %&gt;% 
  dplyr::count(cell_type) %&gt;% 
  ggplot(.,aes(reorder(cell_type,n),n,fill=cell_type)) + 
  geom_col() + 
  coord_flip() + 
  theme_classic() + 
  theme(legend.position = &#39;none&#39;,text=element_text(size=20)) + 
  ylab(&#39;Number of cis-eQTL (5% FDR)&#39;) + xlab(&#39;&#39;)
ggsave(p,filename=&#39;output/figures/Figure2A.png&#39;,height=5,width=7,dpi = 300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-2b" class="section level2">
<h2>Figure 2B</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;) %&gt;% 
  filter(adj_p&lt;0.05) %&gt;% 
  dplyr::count(cell_type)</code></pre>
<pre class="r"><code>sum_expression &lt;- readRDS(&#39;data_sensitive/expression/all_sum_expression.rds&#39;)</code></pre>
<pre class="r"><code>n_cells_tot &lt;- sum_expression %&gt;% 
  dplyr::select(cell_type,individual_id,n_cells) %&gt;% 
  unique() %&gt;% 
  group_by(cell_type) %&gt;% 
  summarise(`Number of cells`=sum(n_cells))</code></pre>
<pre class="r"><code>d &lt;- left_join(d,n_cells_tot,by=&#39;cell_type&#39;)</code></pre>
<pre class="r"><code>p &lt;- ggplot(d,aes(`Number of cells`,n,col=cell_type)) + 
  geom_point() + 
  ggrepel::geom_label_repel(aes(label=cell_type),size=5,box.padding = 0.5) + 
  theme_classic() + 
  theme(legend.position = &#39;none&#39;,text=element_text(size=20)) + 
  scale_x_log10() + 
  scale_y_log10() + 
  ylab(&#39;Number of cis-eQTL (5%FDR)&#39;)
ggsave(p,filename=&#39;output/figures/Figure2B.png&#39;,height=5,width=7,dpi = 300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-2c" class="section level2">
<h2>Figure 2C</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;) %&gt;% 
  filter(adj_p&lt;0.05) %&gt;% 
  dplyr::select(cell_type,pid,sid) %&gt;% 
  dplyr::rename(SNP=sid) %&gt;% 
  separate(pid,into=c(&#39;symbol&#39;,&#39;ensembl&#39;),sep=&#39;_&#39;)</code></pre>
<pre class="bash"><code>cd data_sensitive/genotypes/processed
zcat combined_final.vcf.gz | cut -f 1,2,3 | grep -v &quot;^##&quot; &gt; snp_pos_hg38.txt</code></pre>
<pre class="r"><code>snp_pos &lt;- data.table::fread(&#39;data_sensitive/genotypes/processed/snp_pos_hg38.txt&#39;,data.table = FALSE) %&gt;% 
  dplyr::select(chr_hg38=V1,pos_hg38=V2,SNP=V3) %&gt;% 
  as_tibble()</code></pre>
<pre class="r"><code>d &lt;- left_join(d,snp_pos,by=&#39;SNP&#39;) %&gt;% 
  mutate(SNP_id_hg38=paste0(&#39;chr&#39;,chr_hg38,&#39;:&#39;,pos_hg38))</code></pre>
<pre class="r"><code>get_snp_pvalues_metabrain &lt;- function(chr,tissue=&#39;Cortex&#39;) {

  d_sig_chr &lt;- filter(d,chr_hg38==chr)
  
  matched_pvalues &lt;- 
    data.table::fread(paste0(&#39;data/metabrain/2020-05-26-&#39;,tissue,&#39;-EUR-&#39;,chr,&#39;-biogenformat.txt.gz&#39;),data.table = FALSE) %&gt;% 
    dplyr::select(PValue,SNPChr,SNPChrPos,ProbeName,SNPName) %&gt;% 
    mutate(ensembl=gsub(&#39;\\..+&#39;,&#39;&#39;,ProbeName)) %&gt;% 
    mutate(SNP_id_hg38=paste0(&#39;chr&#39;,SNPChr,&#39;:&#39;,SNPChrPos)) %&gt;% 
    as_tibble() %&gt;% 
    mutate(tissue=tissue) %&gt;% 
    dplyr::select(SNP_id_hg38,ensembl,PValue) %&gt;% 
    inner_join(.,d_sig_chr,by=c(&#39;SNP_id_hg38&#39;,&#39;ensembl&#39;))
  return(matched_pvalues)
}</code></pre>
<pre class="r"><code>matched_snps &lt;- mclapply(1:22,get_snp_pvalues_metabrain,mc.cores=11,mc.preschedule = FALSE) %&gt;% bind_rows()</code></pre>
<pre class="r"><code>write_tsv(matched_snps,&#39;output/eqtl/eqtl.PC70.metabrain.matched.txt&#39;)</code></pre>
<pre class="r"><code>matched_snps &lt;- matched_snps %&gt;% mutate(glia_neurons=case_when(
  cell_type %in% c(&#39;Excitatory neurons&#39;,&#39;Inhibitory neurons&#39;) ~ &#39;Neurons&#39;,
  TRUE ~ &#39;Glia&#39;
)) </code></pre>
<pre class="r"><code>matched_snps_text_prop &lt;- matched_snps %&gt;% 
  group_by(glia_neurons) %&gt;% 
  summarise(prop=paste0(&#39;prop = &#39;,round(sum(PValue&lt;0.05)*100/sum(PValue&lt;10),digits=1),&#39;%&#39;),
            pi1=paste0(&#39;pi1 = &#39;,round(1-qvalue::qvalue(PValue)$pi0,digits=2)),
            label=paste0(prop,&#39;\n&#39;,pi1),n_eqtl=sum(PValue&lt;10))

p &lt;- ggplot(matched_snps,aes(PValue,fill=glia_neurons)) + 
  geom_histogram() + 
  facet_wrap(~glia_neurons,scales=&#39;free_y&#39;) + 
  theme_classic() + 
  theme(legend.position = &#39;none&#39;,text=element_text(size=20)) +
  geom_text(data=matched_snps_text_prop,aes(x=0.5,y=Inf,vjust=2.1,label=label),size=8) +
  ggtitle(&#39;Metabrain Cortex matched pvalues&#39;)
ggsave(p,filename=&#39;output/figures/Figure2C.png&#39;,height=5,width=7,dpi = 300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-26-1">
Past versions of unnamed-chunk-26-1.png
</button>
</p>
<div id="fig-unnamed-chunk-26-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
056f5c7
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
</tr>
<tr>
<td>
811141f
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="figure-2d" class="section level2">
<h2>Figure 2D</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;) %&gt;% 
  filter(adj_p&lt;0.05)

p &lt;- d %&gt;% 
  ggplot(aes(dist,fill=cell_type)) + 
  geom_histogram(bins = 50) + 
  facet_wrap(~cell_type,scales = &#39;free_y&#39;,nrow=4,ncol=2) + 
  theme_classic() + 
  theme(legend.position = &#39;none&#39;,
        text=element_text(size=20),
        plot.margin =unit(c(5.5, 20, 5.5, 5.5), &quot;points&quot;)) + 
  xlab(&#39;Distance to TSS&#39;)
ggsave(p,filename=&#39;output/figures/Figure2D.png&#39;,height=6,width=9,dpi = 300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-27-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-27-1">
Past versions of unnamed-chunk-27-1.png
</button>
</p>
<div id="fig-unnamed-chunk-27-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
056f5c7
</td>
<td>
Julien Bryois
</td>
<td>
2021-12-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="figure-2e" class="section level2">
<h2>Figure 2E</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;)</code></pre>
<pre class="r"><code>loeuf &lt;- read_tsv(&#39;data/gnomad_loeuf/supplementary_dataset_11_full_constraint_metrics.tsv&#39;) %&gt;% 
  filter(canonical==TRUE) %&gt;% 
  dplyr::select(gene,gene_id,transcript,oe_lof_upper,p) %&gt;% 
  mutate(oe_lof_upper_bin = ntile(oe_lof_upper, 10)) %&gt;% 
  dplyr::rename(ensembl=gene_id) %&gt;% 
  dplyr::select(-gene,-transcript)</code></pre>
<pre class="r"><code>d_loeuf &lt;- d %&gt;% 
  separate(pid,into=c(&#39;symbol&#39;,&#39;ensembl&#39;),sep=&#39;_&#39;) %&gt;% 
  left_join(.,loeuf,by=&#39;ensembl&#39;) %&gt;% 
  mutate(Significant=ifelse(adj_p&lt;0.05,TRUE,FALSE)) %&gt;% 
  filter(!is.na(oe_lof_upper))</code></pre>
<pre class="r"><code>my_comparisons &lt;- list( c(&quot;FALSE&quot;,&quot;TRUE&quot;))
p &lt;- ggplot(d_loeuf,aes(Significant,oe_lof_upper,col=cell_type)) + ggbeeswarm::geom_quasirandom(size=0.1) + facet_wrap(~cell_type,ncol=4,nrow=2) + theme_classic() + theme(legend.position=&#39;none&#39;,text=element_text(size=20)) + stat_summary(fun = median, fun.min = median, fun.max = median,
                 geom = &quot;crossbar&quot;, width = 0.5,size=0.25,color=&#39;black&#39;) + xlab(&quot;Significant (5% FDR)&quot;) + ylab(&quot;LoF observed/expected upper bound
fraction (LOEUF)&quot;) + ggpubr::stat_compare_means(comparisons = my_comparisons) + scale_y_continuous(expand=c(0.1,0))
ggsave(p,filename=&#39;output/figures/Figure2E.png&#39;,width = 9,height=6,dpi = 300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-31-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-2f" class="section level2">
<h2>Figure 2F</h2>
<pre class="r"><code>hdF5_file_path &lt;- &#39;output/shiny/data.h5&#39;</code></pre>
<pre class="r"><code>eqtl &lt;- h5read(hdF5_file_path, &quot;eqtl_results/eqtl_results_all&quot;)</code></pre>
<pre class="r"><code>plot_eqtl &lt;- function(cell_type_name,gene_name,snp_name,col_point){
  
  gene_short &lt;- gsub(&#39;_.+&#39;,&#39;&#39;,gene_name)
  
  genotype &lt;- h5read(hdF5_file_path, paste0(&quot;genotype/&quot;,snp_name))
  expression &lt;- h5read(hdF5_file_path, paste0(&quot;expression/&quot;,gene_name)) %&gt;% 
    dplyr::rename(individual=individual_id)
  
  pval &lt;- filter(eqtl,gene==gene_name,cell_type==cell_type_name) %&gt;% pull(adj_p)
  
  d &lt;- inner_join(genotype,expression,by=&#39;individual&#39;) %&gt;% 
  mutate(genotype_label = factor(case_when(
    genotype == 0 ~ paste0(REF,REF),
    genotype == 1 ~ paste0(REF,ALT),
    genotype == 2 ~ paste0(ALT,ALT)
  ),levels=c(unique(paste0(REF,REF)),unique(paste0(REF,ALT)),unique(paste0(ALT,ALT))))) %&gt;% filter(cell_type%in%cell_type_name)
  
  #Only display points for genotypes with more than one individual
  n_genotypes &lt;- d %&gt;% dplyr::count(genotype_label) %&gt;% 
    filter(n&gt;1)
  
  d &lt;- filter(d,genotype_label%in%n_genotypes$genotype_label)
  
  p &lt;- ggplot(d, aes(genotype_label,log2_cpm)) + 
    ggbeeswarm::geom_quasirandom(size=2,col=col_point) + 
      geom_boxplot(alpha=0.05,aes(group=genotype_label),outlier.shape = NA,col=col_point) + 
    theme_classic() + 
    theme(legend.position=&#39;none&#39;,text=element_text(size=16), plot.title = element_text(size=10)) + 
    xlab(&#39;Genotype&#39;) + 
    ylab(&#39;Expression (cpm) (log2+1)&#39;) +
    ggtitle(paste0(gene_short,&#39; - &#39;,snp_name,&#39; - &#39;,cell_type_name,&#39;\nadjusted p-value =&#39;,signif(pval,digits=2)))
  p
}</code></pre>
<pre class="r"><code>gene_name &lt;- &#39;HLA-DRB5_ENSG00000198502&#39;
snp_name &lt;- &#39;rs9272114&#39;
cell_type_name &lt;- &#39;Microglia&#39;</code></pre>
<pre class="r"><code>hla &lt;- plot_eqtl(cell_type_name,gene_name,snp_name,&#39;#55bdc2&#39;) + theme(axis.title.y = element_blank())</code></pre>
<pre class="r"><code>gene_name &lt;- &#39;NSUN2_ENSG00000037474&#39;
snp_name &lt;- &#39;rs567289&#39;
cell_type_name &lt;- &#39;Astrocytes&#39;</code></pre>
<pre class="r"><code>nsun2 &lt;- plot_eqtl(cell_type_name,gene_name,snp_name,&#39;#e97e72&#39;)</code></pre>
<pre class="r"><code>p &lt;- nsun2 + hla &amp; theme(text=element_text(size=24),plot.title = element_text(size=20))
p &amp; theme(text=element_text(size=12),plot.title = element_text(size=12))</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-39-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(p,filename = &#39;output/figures/Figure2F.png&#39;,width=12,height=5)</code></pre>
</div>
<div id="figure-2g" class="section level2">
<h2>Figure 2G</h2>
<pre class="r"><code>gene_name &lt;- &#39;MET_ENSG00000105976&#39;
snp_name &lt;- &#39;rs187437&#39;
cell_type_name &lt;- &#39;Oligodendrocytes&#39;</code></pre>
<pre class="r"><code>met &lt;- plot_eqtl(cell_type_name,gene_name,snp_name,&#39;#48a8f8&#39;)</code></pre>
<pre class="r"><code>gene_name &lt;- &#39;GLUD1_ENSG00000148672&#39;
snp_name &lt;- &#39;rs17096421&#39;
cell_type_name &lt;- &#39;Excitatory neurons&#39;</code></pre>
<pre class="r"><code>glud1 &lt;- plot_eqtl(cell_type_name,gene_name,snp_name,&#39;#87ab34&#39;) + theme(axis.title.y = element_blank())</code></pre>
<pre class="r"><code>p &lt;- met + glud1 &amp; theme(text=element_text(size=24),plot.title = element_text(size=20))
p &amp; theme(text=element_text(size=12),plot.title = element_text(size=12))</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-44-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(p,filename = &#39;output/figures/Figure2E.png&#39;,width=12,height=5)</code></pre>
</div>
<div id="figure-3a" class="section level2">
<h2>Figure 3A</h2>
<pre class="r"><code>fdensity_files &lt;- list.files(&#39;data/epigenome_enrichment/fdensity/results/corces/&#39;,full.names = TRUE)</code></pre>
<pre class="r"><code>fdensity_res &lt;- tibble(fdensity_files) %&gt;% mutate(file_content=map(fdensity_files,read_delim,delim=&#39; &#39;,col_names=FALSE)) %&gt;% 
  unnest(file_content) %&gt;% 
  mutate(name=basename(fdensity_files)) %&gt;% 
  dplyr::select(-fdensity_files) %&gt;% 
  separate(name,into=c(&#39;tmp&#39;,&#39;sig&#39;,&#39;annot&#39;),sep=&#39;_&#39;) %&gt;% 
  dplyr::select(-tmp) %&gt;% 
  mutate(sig=gsub(&#39;.sig5FDR.bed&#39;,&#39;&#39;,sig)) %&gt;% 
  mutate(annot=gsub(&#39;.bed&#39;,&#39;&#39;,annot)) %&gt;% 
  group_by(sig,annot) %&gt;% 
  mutate(X3_norm=X3/sum(X3)) %&gt;% 
  mutate(sig=gsub(&#39;\\.&#39;,&#39; &#39;,sig) %&gt;% gsub(&#39;OPCs   COPs&#39;,&#39;OPCs / COPs&#39;,.)) %&gt;% 
  mutate(annot=gsub(&#39;\\.&#39;,&#39; &#39;,annot)) %&gt;% 
  mutate(annot=gsub(&#39; corces&#39;,&#39;&#39;,annot)) %&gt;% 
  mutate(annot=gsub(&#39;ExcitatoryNeurons&#39;,&#39;Excitatory neurons&#39;,annot)) %&gt;% 
  mutate(annot=gsub(&#39;InhibitoryNeurons&#39;,&#39;Inhibitory neurons&#39;,annot)) %&gt;% 
  mutate(annot=gsub(&#39;OPCs&#39;,&#39;OPCs / COPs&#39;,annot)) %&gt;% 
  mutate(annot=gsub(&#39;specific&#39;,&#39;ATAC (specific)&#39;,annot)) %&gt;% 
  mutate(sig=paste0(sig,&#39; eQTL&#39;)) %&gt;% 
  mutate(sig=factor(case_when(
    sig==&#39;Astrocytes eQTL&#39; ~ &#39;Astro eQTL&#39;,
    sig==&#39;Endothelial cells eQTL&#39; ~ &#39;Endo eQTL&#39;,
    sig==&#39;Excitatory neurons eQTL&#39; ~ &#39;Ex eQTL&#39;,
    sig==&#39;Inhibitory neurons eQTL&#39; ~ &#39;Inh eQTL&#39;,
    sig==&#39;Microglia eQTL&#39; ~ &#39;Micro eQTL&#39;,
    sig==&#39;Oligodendrocytes eQTL&#39; ~ &#39;Oligo eQTL&#39;,
    sig==&#39;OPCs / COPs eQTL&#39; ~ &#39;OPCs eQTL&#39;,
    sig==&#39;Pericytes eQTL&#39; ~ &#39;Pericytes eQTL&#39;),
    levels=c(&#39;Astro eQTL&#39;,&#39;Micro eQTL&#39;,&#39;Oligo eQTL&#39;,&#39;OPCs eQTL&#39;,&#39;Ex eQTL&#39;,&#39;Inh eQTL&#39;,&#39;Endo eQTL&#39;,&#39;Pericytes eQTL&#39;))) %&gt;% 
  filter(grepl(&#39;ATAC&#39;,annot)) %&gt;% 
  mutate(annot=factor(case_when(
    annot==&#39;Astrocytes ATAC (specific)&#39; ~ &#39;Astro ATAC&#39;,
    annot==&#39;Excitatory neurons ATAC (specific)&#39; ~ &#39;Ex ATAC&#39;,
    annot==&#39;Inhibitory neurons ATAC (specific)&#39; ~ &#39;Inh ATAC&#39;,
    annot==&#39;Microglia ATAC (specific)&#39; ~ &#39;Micro ATAC&#39;,
    annot==&#39;Oligodendrocytes ATAC (specific)&#39; ~ &#39;Oligo ATAC&#39;,
    annot==&#39;OPCs / COPs ATAC (specific)&#39; ~ &#39;OPCs ATAC&#39;),levels=c(&#39;Astro ATAC&#39;,&#39;Micro ATAC&#39;,&#39;Oligo ATAC&#39;,&#39;OPCs ATAC&#39;,&#39;Ex ATAC&#39;,&#39;Inh ATAC&#39;)))</code></pre>
<pre class="r"><code>cols &lt;- scales::hue_pal()(8)[c(1,5,6,7,3,4)]
p &lt;- fdensity_res  %&gt;% filter(!sig%in%c(&#39;Pericytes eQTL&#39;,&#39;Endo eQTL&#39;),grepl(&#39;ATAC&#39;,annot)) %&gt;% ggplot(.,aes((X1+X2)/2,X3_norm,col=sig)) + geom_line() + facet_grid(sig~annot) + theme_classic() + theme(legend.position = &#39;none&#39;,text=element_text(size=18),axis.title = element_text(size=30)) + xlab(&#39;Distance to cis-eQTL (kb)&#39;) + ylab(&#39;Density&#39;) + scale_x_continuous(breaks=c(-500000,0,500000),labels = c(&#39;-500&#39;,&#39;0&#39;,&#39;500&#39;)) + scale_y_continuous(breaks=c(0.005,0.010,0.015),labels = c(&#39;0.005&#39;,&#39;0.01&#39;,&#39;0.015&#39;)) + scale_color_manual(values=cols)
ggsave(p,filename = &#39;output/figures/Figure3A.png&#39;,width=11,height=8.5,dpi = 300)
p + theme(text=element_text(size=10),axis.title = element_text(size=14))</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-47-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-3b" class="section level2">
<h2>Figure 3B</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl_specific/eqtl.PC70.specific.txt&#39;) %&gt;% 
  #Sets pvalue to NA if the model did not converge
  mutate(nb_pvalue_aggregate=ifelse(nb_pvalue_aggregate_model_converged==FALSE,NA,nb_pvalue_aggregate)) %&gt;% 
  mutate(nb_pvalue_at_least_one=ifelse(nb_pvalue_at_least_one_model_converged==FALSE,NA,nb_pvalue_at_least_one)) %&gt;%   #Remove genes for which the model did not converge (9 genes)
  filter(!is.na(nb_pvalue_aggregate),
         !is.na(nb_pvalue_at_least_one),
         nb_pvalue_at_least_one!=Inf) %&gt;% 
  #Get adjusted pvalues
  mutate(nb_pvalue_aggregate_adj=p.adjust(nb_pvalue_aggregate,method=&#39;fdr&#39;),
         nb_pvalue_at_least_one_adj=p.adjust(nb_pvalue_at_least_one,method = &#39;fdr&#39;)) %&gt;% 
  #For each row, get the maximum pvalue across all cell types, this will be the gene-level pvalue testing whether the genetic effect on gene expression is different than all other cell types
  rowwise() %&gt;% 
  mutate(nb_pvalue_sig_all=max(Astrocytes_p,`Endothelial cells_p`,`Excitatory neurons_p`,`Inhibitory neurons_p`,Microglia_p,Oligodendrocytes_p,`OPCs / COPs_p`,Pericytes_p,na.rm=TRUE)) %&gt;% 
  ungroup() %&gt;% 
  mutate(nb_pvalue_all_adj = p.adjust(nb_pvalue_sig_all,method=&#39;fdr&#39;))</code></pre>
<pre class="r"><code>d_short &lt;- d %&gt;% 
  dplyr::select(cell_type_id,gene_id,snp_id,nb_pvalue_aggregate_adj,nb_pvalue_at_least_one_adj,nb_pvalue_all_adj)</code></pre>
<pre class="r"><code>p &lt;- d_short %&gt;% ungroup() %&gt;% 
  group_by(cell_type_id) %&gt;% 
  summarise(
  `Cell type specific (aggregate)`=sum(nb_pvalue_aggregate_adj&lt;0.05,na.rm=TRUE),
  `Cell type specific (at least one)`=sum(nb_pvalue_at_least_one_adj&lt;0.05,na.rm=TRUE),
  `Cell type specific (all)`=sum(nb_pvalue_all_adj&lt;0.05),
  `Total eQTLs (converged)`=n()) %&gt;% 
  tidyr::gather(type,value,-cell_type_id) %&gt;% 
  mutate(type=factor(type,levels=rev(c(&#39;Total eQTLs (converged)&#39;,&#39;Cell type specific (at least one)&#39;,&#39;Cell type specific (aggregate)&#39;,&#39;Cell type specific (all)&#39;)))) %&gt;% 
  mutate(cell_type_id=factor(cell_type_id,levels=rev(c(&#39;Excitatory neurons&#39;,&#39;Oligodendrocytes&#39;,&#39;Astrocytes&#39;,&#39;Inhibitory neurons&#39;,&#39;OPCs / COPs&#39;,&#39;Microglia&#39;,&#39;Endothelial cells&#39;,&#39;Pericytes&#39;)))) %&gt;% 
  ggplot(.,aes(value,cell_type_id,fill=type)) +
  geom_col(position = position_dodge()) + 
  guides(fill = guide_legend(reverse = TRUE)) +
  geom_text(aes(label=value),position = position_dodge(width = 0.9),hjust=-0.1) +
  xlim(c(0,2950)) + 
  ylab(&#39;&#39;) + 
  theme_classic() + 
  xlab(&#39;Number of eQTL (5%FDR)&#39;) + 
  scale_fill_hue(direction = -1) + 
  theme(legend.position=&#39;right&#39;, text=element_text(size=18))
ggsave(p,filename=&#39;output/figures/Figure3B.png&#39;,height=6,width=10,dpi = 300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-50-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-3c" class="section level2">
<h2>Figure 3C</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl_specific/eqtl.PC70.specific.txt&#39;) %&gt;% 
  #Sets pvalue to NA if the model did not converge
  mutate(nb_pvalue_aggregate=ifelse(nb_pvalue_aggregate_model_converged==FALSE,NA,nb_pvalue_aggregate)) %&gt;% 
  mutate(nb_pvalue_at_least_one=ifelse(nb_pvalue_at_least_one_model_converged==FALSE,NA,nb_pvalue_at_least_one)) %&gt;%   #Remove genes for which the model did not converge (9 genes)
  filter(!is.na(nb_pvalue_aggregate),
         !is.na(nb_pvalue_at_least_one),
         nb_pvalue_at_least_one!=Inf)</code></pre>
<pre class="r"><code>d_gather &lt;- d %&gt;% 
  dplyr::select(-snp_id) %&gt;% 
  gather(cell_type_test,p,-cell_type_id,-gene_id,-nb_pvalue_aggregate,-nb_pvalue_at_least_one,-nb_pvalue_aggregate_model_converged,-nb_pvalue_at_least_one_model_converged) %&gt;% 
  mutate(cell_type_test=gsub(&#39;_p&#39;,&#39;&#39;,cell_type_test)) %&gt;% 
  mutate(cell_type_id=factor(cell_type_id,levels=c(&#39;Microglia&#39;,&#39;OPCs / COPs&#39;, &#39;Astrocytes&#39;,&#39;Oligodendrocytes&#39;,&#39;Inhibitory neurons&#39;, &#39;Excitatory neurons&#39;,&#39;Endothelial cells&#39;,&#39;Pericytes&#39;)),
         cell_type_test=factor(cell_type_test,levels=c(&#39;Microglia&#39;,&#39;OPCs / COPs&#39;, &#39;Astrocytes&#39;,&#39;Oligodendrocytes&#39;,&#39;Inhibitory neurons&#39;, &#39;Excitatory neurons&#39;,&#39;Endothelial cells&#39;,&#39;Pericytes&#39;)))</code></pre>
<pre class="r"><code>pi1 &lt;- d_gather %&gt;% 
  filter(!is.na(p)) %&gt;% 
  filter(!cell_type_id%in%c(&#39;Endothelial cells&#39;,&#39;Pericytes&#39;)) %&gt;% 
  group_by(cell_type_id,cell_type_test) %&gt;% 
  summarise(pi1=1-qvalue::qvalue(p)$pi0)</code></pre>
<pre class="r"><code>pi1 %&gt;% spread(cell_type_test,pi1) %&gt;% column_to_rownames(&#39;cell_type_id&#39;) %&gt;% pheatmap::pheatmap(.,display_numbers = TRUE,number_color = &#39;black&#39;,angle_col = 45) %&gt;% print()</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-54-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pi1 %&gt;% spread(cell_type_test,pi1) %&gt;% column_to_rownames(&#39;cell_type_id&#39;) %&gt;% pheatmap::pheatmap(.,display_numbers = TRUE,number_color = &#39;black&#39;,angle_col = 45,filename = &#39;output/figures/Figure3C.png&#39;,height = 4)</code></pre>
</div>
<div id="figure-3d" class="section level2">
<h2>Figure 3D</h2>
<pre class="r"><code>hdF5_file_path &lt;- &#39;output/shiny/data.h5&#39;</code></pre>
<pre class="r"><code>eqtl_specific &lt;- h5read(hdF5_file_path, &quot;eqtl_results/eqtl_results_specific&quot;)</code></pre>
<pre class="r"><code>plot_eqtl_specific &lt;- function(cell_type_name,gene_name,snp_name){
  
  gene_short &lt;- gsub(&#39;_.+&#39;,&#39;&#39;,gene_name)
  
  genotype &lt;- h5read(hdF5_file_path, paste0(&quot;genotype/&quot;,snp_name))
  expression &lt;- h5read(hdF5_file_path, paste0(&quot;expression/&quot;,gene_name)) %&gt;% 
    dplyr::rename(individual=individual_id)
  
  pval &lt;- filter(eqtl_specific,gene==gene_name,cell_type==cell_type_name) %&gt;% pull(nb_pvalue_all_adj)
  
  d &lt;- inner_join(genotype,expression,by=&#39;individual&#39;) %&gt;% 
  mutate(genotype_label = case_when(
    genotype == 0 ~ paste0(REF,REF),
    genotype == 1 ~ paste0(REF,ALT),
    genotype == 2 ~ paste0(ALT,ALT)
  )) %&gt;%  mutate(cell_type=fct_relevel(cell_type,cell_type_name))
  
  p &lt;- ggplot(d, aes(genotype_label,log2_cpm,col=cell_type)) + 
    ggbeeswarm::geom_quasirandom(size=0.5) + 
      geom_boxplot(alpha=0.05,aes(group=genotype_label),outlier.shape = NA) + 
    theme_classic() + 
    theme(legend.position=&#39;none&#39;,text=element_text(size=16), plot.title = element_text(size=10)) + 
    xlab(&#39;Genotype&#39;) + 
    ylab(&#39;Expression (cpm) (log2+1)&#39;) +
    ggtitle(paste0(gene_short,&#39; - &#39;,snp_name,&#39; - &#39;,cell_type_name,&#39;\nadjusted pvalue = &#39;, pval)) + facet_wrap(~cell_type,ncol=4) + scale_color_discrete(limits=unique(d$cell_type) %&gt;% as.character() %&gt;% sort())
  p
}</code></pre>
<pre class="r"><code>gene_name &lt;- &#39;CHRM5_ENSG00000184984&#39;
snp_name &lt;- &#39;rs8038713&#39;
cell_type_name &lt;- &#39;Oligodendrocytes&#39;
chrm5 &lt;- plot_eqtl_specific(cell_type_name,gene_name,snp_name) + theme(text=element_text(size=20),plot.title = element_text(size=12))
chrm5 + theme(text=element_text(size=14))</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-59-1.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(chrm5,filename = &#39;output/figures/Figure3D.png&#39;,width=9,height=5)</code></pre>
</div>
<div id="figure-3e" class="section level2">
<h2>Figure 3E</h2>
<pre class="r"><code>gene_name &lt;- &#39;RNF150_ENSG00000170153&#39;
snp_name &lt;- &#39;rs13131368&#39;
cell_type_name &lt;- &#39;Microglia&#39;
p &lt;- plot_eqtl_specific(cell_type_name,gene_name,snp_name) + theme(text=element_text(size=20),plot.title = element_text(size=12))
p + theme(text=element_text(size=14))</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-60-1.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(p,filename = &#39;output/figures/Figure3E.png&#39;,width=9,height=5)</code></pre>
</div>
<div id="figure-3f" class="section level2">
<h2>Figure 3F</h2>
<pre class="r"><code>gene_name &lt;- &#39;ACAP2_ENSG00000114331&#39;
snp_name &lt;- &#39;rs6437374&#39;
cell_type_name &lt;- &#39;Astrocytes&#39;
p &lt;- plot_eqtl_specific(cell_type_name,gene_name,snp_name) + theme(text=element_text(size=20),plot.title = element_text(size=12))
p + theme(text=element_text(size=14))</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-61-1.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(p,filename = &#39;output/figures/Figure3F.png&#39;,width=9,height=5)</code></pre>
</div>
<div id="figure-4a" class="section level2">
<h2>Figure 4A</h2>
<pre class="r"><code>read_coloc &lt;- function(path){
  coloc &lt;- read_tsv(path) %&gt;% 
    dplyr::select(locus,closest_gene,ensembl,symbol,tissue,PP.H4.abf) %&gt;% 
    dplyr::rename(phenotype=ensembl,hgnc_symbol=symbol,locus_name=locus)
  
  #If coloc was performed for the same genes located in loci in close proximity, take the best
  coloc &lt;- coloc %&gt;% group_by(phenotype,tissue) %&gt;% 
  slice_max(n=1,order_by=PP.H4.abf,with_ties=FALSE) %&gt;% 
  ungroup()
  
  return(coloc)
}</code></pre>
<pre class="r"><code>coloc_ms &lt;- read_coloc(&#39;output/coloc/coloc.ms.txt&#39;) %&gt;% mutate(trait=&#39;Multiple Sclerosis&#39;) 
coloc_ad &lt;- read_coloc(&#39;output/coloc/coloc.ad.txt&#39;) %&gt;% mutate(trait=&#39;Alzheimer&#39;)
coloc_scz &lt;- read_coloc(&#39;output/coloc/coloc.scz.txt&#39;) %&gt;% mutate(trait=&#39;Schizophrenia&#39;)
coloc_pd &lt;- read_coloc(&#39;output/coloc/coloc.pd.txt&#39;) %&gt;% mutate(trait=&#39;Parkinson&#39;)</code></pre>
<pre class="r"><code>coloc &lt;- rbind(coloc_ms,coloc_ad,coloc_scz,coloc_pd)</code></pre>
<pre class="r"><code>my_ceil &lt;- function(x) {
  ceil &lt;- ceiling(max(x))
  #ifelse(ceil &gt; 1 &amp; ceil %% 2 == 1, ceil + 1, ceil)
}

my_breaks &lt;- function(x) { 
  ceil &lt;- my_ceil(max(x))
  unique(ceiling(pretty(seq(0, ceil))))
} 

my_limits &lt;- function(x) { 
  ceil &lt;- my_ceil(x[2])
  c(x[1], ceil)
}</code></pre>
<pre class="r"><code>x &lt;- coloc %&gt;% 
  filter(PP.H4.abf&gt;=0.7) %&gt;% 
  dplyr::select(trait,closest_gene,hgnc_symbol,tissue) %&gt;% 
  group_by(trait,closest_gene,hgnc_symbol) %&gt;% 
  mutate(n_cell_types_per_locus=length(unique(tissue))) %&gt;% 
  ungroup() %&gt;% 
  group_by(trait,closest_gene) %&gt;% 
  mutate(n_genes_per_loci=length(unique(hgnc_symbol))) %&gt;% 
  ungroup() %&gt;% 
  mutate(cell_type=ifelse(n_cell_types_per_locus&gt;1,&#39;Multiple&#39;,tissue)) %&gt;% 
  dplyr::select(-n_cell_types_per_locus,-tissue) %&gt;% 
  unique() %&gt;% 
  group_by(closest_gene) %&gt;% 
  mutate(n_cell_type=length(unique(cell_type))) %&gt;% 
  ungroup() %&gt;% 
  mutate(cell_type2=ifelse(n_cell_type&gt;1,&#39;Multiple&#39;,cell_type)) %&gt;%
  dplyr::select(trait,closest_gene,hgnc_symbol,n_genes_per_loci,cell_type2) %&gt;% 
  mutate(trait=factor(trait,levels=c(&#39;Alzheimer&#39;,&#39;Parkinson&#39;,&#39;Multiple Sclerosis&#39;,&#39;Schizophrenia&#39;))) %&gt;%
  dplyr::select(trait,closest_gene,n_genes_per_loci,cell_type2) %&gt;% unique() %&gt;% 
  dplyr::count(trait,n_genes_per_loci,cell_type2)</code></pre>
<pre class="r"><code>p &lt;-  ggplot(x,aes(n_genes_per_loci,n,fill=cell_type2)) + geom_col() + facet_wrap(~trait,scales = &#39;free&#39;) + xlab(&#39;Number of coloc genes&#39;) + ylab(&#39;Number of loci&#39;) + theme_classic() + scale_x_continuous(breaks=my_breaks,limits=my_limits) + guides(fill=guide_legend(title=&#39;&#39;))
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-67-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>p &lt;- p + theme(text=element_text(size=50),axis.ticks.length=unit(.25, &quot;cm&quot;),legend.position=&#39;top&#39;,legend.text = element_text(size=22)) 
ggsave(p,filename=&#39;output/figures/Figure4A.png&#39;,width=16,height=12,dpi = 300)</code></pre>
</div>
<div id="figure-4b" class="section level2">
<h2>Figure 4B</h2>
<pre class="r"><code>read_coloc &lt;- function(path){
  coloc &lt;- read_tsv(path) %&gt;% 
    dplyr::select(locus,closest_gene,ensembl,symbol,tissue,PP.H4.abf) %&gt;% 
    dplyr::rename(phenotype=ensembl,hgnc_symbol=symbol,locus_name=locus)
  
  #If coloc was performed for the same genes located in loci in close proximity, take the best
  coloc &lt;- coloc %&gt;% group_by(phenotype,tissue) %&gt;% 
  slice_max(n=1,order_by=PP.H4.abf,with_ties=FALSE) %&gt;% 
  ungroup()
  
  return(coloc)
}</code></pre>
<pre class="r"><code>coloc_ms &lt;- read_coloc(&#39;output/coloc/coloc.ms.txt&#39;) %&gt;% mutate(trait=&#39;Multiple Sclerosis&#39;) 
coloc_ad &lt;- read_coloc(&#39;output/coloc/coloc.ad.txt&#39;) %&gt;% mutate(trait=&#39;Alzheimer&#39;)
coloc_scz &lt;- read_coloc(&#39;output/coloc/coloc.scz.txt&#39;) %&gt;% mutate(trait=&#39;Schizophrenia&#39;)
coloc_pd &lt;- read_coloc(&#39;output/coloc/coloc.pd.txt&#39;) %&gt;% mutate(trait=&#39;Parkinson&#39;)</code></pre>
<pre class="r"><code>coloc &lt;- rbind(coloc_ms,coloc_ad,coloc_scz,coloc_pd)</code></pre>
<pre class="r"><code>genes_with_coloc &lt;- coloc %&gt;% 
  group_by(trait) %&gt;% 
  filter(PP.H4.abf&gt;0.7) %&gt;% 
  dplyr::select(phenotype,trait) %&gt;% 
  unique()</code></pre>
<pre class="r"><code>coloc_selected &lt;- coloc %&gt;% 
  inner_join(.,genes_with_coloc,by=c(&#39;phenotype&#39;,&#39;trait&#39;)) %&gt;% 
  group_by(phenotype,trait) %&gt;% 
  mutate(rank_coloc=rank(-PP.H4.abf)) %&gt;% 
  mutate(lab=case_when(
    hgnc_symbol==&#39;BIN1&#39; ~ &#39;BIN1&#39;,
    hgnc_symbol==&#39;TMEM163&#39; ~ &#39;TMEM163&#39;,
    hgnc_symbol==&#39;NR1H3&#39; ~ &#39;NR1H3&#39;,
    hgnc_symbol==&#39;RERE&#39; ~ &#39;RERE&#39;,
    TRUE ~ &#39;&#39;
  )) %&gt;% 
    mutate(lab_size=case_when(
    hgnc_symbol==&#39;BIN1&#39; ~ 8,
    hgnc_symbol==&#39;TMEM163&#39; ~ 8,
    hgnc_symbol==&#39;NR1H3&#39; ~ 8,
    hgnc_symbol==&#39;RERE&#39; ~ 8,
    TRUE ~ 4
  ))</code></pre>
<pre class="r"><code>my_comparisons &lt;- list( c(&quot;1&quot;, &quot;2&quot;))
coloc_selected &lt;- mutate(coloc_selected,trait=factor(trait,levels=c(&#39;Alzheimer&#39;,&#39;Parkinson&#39;,&#39;Multiple Sclerosis&#39;,&#39;Schizophrenia&#39;)))

p &lt;- ggplot(coloc_selected,aes(as.factor(rank_coloc),PP.H4.abf,group=rank_coloc)) +
  ggbeeswarm::geom_quasirandom(aes(col=lab,size=lab_size,alpha=lab_size)) + 
  geom_boxplot(alpha=0.1,outlier.shape = NA) + 
  facet_wrap(~trait,ncol=2) + 
  theme_classic() + 
  xlab(&#39;Cell type rank&#39;) + 
  ylab(&#39;Posterior\nprobability&#39;)  +
  scale_y_continuous(limits = c(0,1.2),breaks=c(0,0.25,0.5,0.75,1)) +
  scale_x_discrete(expand=c(0.1,0)) + 
  scale_color_manual(values=c(&#39;black&#39;,&#39;#1B9E77&#39;,&#39;#D95F02&#39;,&#39;#7570B3&#39;,&#39;#E7298A&#39;)) +
  guides(size=FALSE) + 
  scale_alpha_binned(range = c(0.3,0.8),guide=&#39;none&#39;) + 
  theme(legend.position=&#39;top&#39;,text=element_text(size=48),legend.title = element_blank()) +
  guides(colour = guide_legend(override.aes = list(size=10))) + 
  scale_size_continuous(range = c(4, 14))

p_figure &lt;- p + ggpubr::stat_compare_means(comparisons = my_comparisons,size=10,vjust = -0.3,bracket.size = 1)

ggsave(p_figure,filename=&#39;output/figures/Figure4B.png&#39;,width=19,height=12,dpi = 300)

p + theme(legend.position=&#39;top&#39;,text=element_text(size=14),legend.title = element_blank()) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  ggpubr::stat_compare_means(comparisons = my_comparisons,size=4,vjust = -0.3,bracket.size = 1)</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-73-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-4c" class="section level2">
<h2>Figure 4C</h2>
<pre class="r"><code>coloc_heatmap &lt;- function(file,threshold=0.7,width=7,height=5,out_name){
    
  coloc_all &lt;- read_tsv(file)

  #Get best coloc if same gene in same cell type was tested in two different loci
  coloc_all &lt;- coloc_all %&gt;% group_by(ensembl,tissue) %&gt;% 
    slice_max(n=1,order_by=PP.H4.abf,with_ties=FALSE) %&gt;% 
    ungroup()
  
  #Get best coloc if same symbol in same cell type was tested multiple times (different ensembl name for same symbol at same locus)
  coloc_all &lt;- coloc_all %&gt;% group_by(symbol,tissue) %&gt;% 
    slice_max(n=1,order_by=PP.H4.abf,with_ties=FALSE) %&gt;% 
    ungroup()

  #Get absolute value of effect size of the GWAS
  coloc_all &lt;- mutate(coloc_all,beta=abs(beta_top_GWAS))
  
  coloc_all_matrix &lt;- dplyr::select(coloc_all,tissue,symbol,PP.H4.abf) %&gt;% 
  filter(!is.na(symbol)) %&gt;% 
  unique() %&gt;% 
  spread(tissue,PP.H4.abf,fill=0) %&gt;% 
  column_to_rownames(&#39;symbol&#39;)

  coloc_all_format_per_gene &lt;- coloc_all %&gt;% 
    group_by(ensembl) %&gt;% 
    filter(PP.H4.abf==max(PP.H4.abf),PP.H4.abf&gt;threshold) %&gt;% 
    mutate(closest_gene=gsub(&#39;ENSG.+ - | - ENSG.+&#39;,&#39;&#39;,closest_gene)) %&gt;% #Remove ENSEMBL name if a symbol + and ensembl gene name
                                                    #are both at same distance from GWAS top pick
    ungroup() %&gt;% 
    dplyr::rename(symbol_coloc=symbol) %&gt;% 
    filter(!is.na(symbol_coloc)) %&gt;% 
    arrange(-PP.H4.abf) %&gt;% 
    mutate(risk=case_when(
      direction ==1 ~ &#39;Up&#39;,
      direction ==-1 ~ &#39;Down&#39;,
      direction ==0 ~ &#39;Isoform&#39;
    )) %&gt;% 
    column_to_rownames(&#39;symbol_coloc&#39;) %&gt;% 
    dplyr::rename(LOEUF=oe_lof_upper_bin)
  
  coloc_all_matrix_subset &lt;- coloc_all_matrix[apply(coloc_all_matrix,1,function(x) any(x&gt;threshold)),] %&gt;% as.matrix()
  
  coloc_all_format_per_gene_direction &lt;- coloc_all_format_per_gene[rownames(coloc_all_matrix_subset),] %&gt;% 
    dplyr::select(closest_gene,risk,LOEUF,beta)
  
 ha = HeatmapAnnotation(df = coloc_all_format_per_gene_direction[,2:3,drop=FALSE],
                         which =&#39;row&#39;,
                         col = list(&#39;risk&#39; = c(&quot;Up&quot;=&quot;#E31A1C&quot;,&quot;Down&quot;=&quot;#1F78B4&quot;,&quot;Isoform&quot;=&quot;darkorange&quot;),
                                    &#39;LOEUF&#39; = circlize::colorRamp2(c(0,10),c(&#39;white&#39;,&#39;springgreen4&#39;))),
                         show_annotation_name = c(&#39;risk&#39; = TRUE,&#39;LOEUF&#39;=TRUE),
                         annotation_name_rot = 45
                         )
  
  hb = HeatmapAnnotation(df = coloc_all_format_per_gene_direction[,4,drop=FALSE],
                         which =&#39;row&#39;,
                         col = list(circlize::colorRamp2(c(min(coloc_all_format_per_gene_direction[[4]],na.rm=T),
                                                                  max(coloc_all_format_per_gene_direction[[4]],na.rm=T)),c(&#39;white&#39;,&#39;palevioletred3&#39;))) %&gt;% 
                           setNames(colnames(coloc_all_format_per_gene_direction)[4]),
                         annotation_name_rot = 45
                         )

   ht &lt;- Heatmap(coloc_all_matrix_subset,col=viridis(100),
                right_annotation=ha,column_names_rot = 45,
                left_annotation = hb,
                #row_names_side = &#39;left&#39;, 
                #row_dend_side = &#39;right&#39;,
                row_split = coloc_all_format_per_gene_direction[, 1],
                row_title_rot = 0,
                cluster_rows = FALSE,
                layer_fun = function(j, i, x, y, width, height, fill) {
                                          grid.text(sprintf(&quot;%.2f&quot;, pindex(coloc_all_matrix_subset, i, j)), 
                                                    x, y, gp = gpar(fontsize = 10,col=&quot;black&quot;))
                                                      },
                #row_title_gp = gpar(col = &#39;red&#39;, font = 2),
                #row_title = rep(&#39;&#39;,length(unique(coloc_all_format_per_gene_direction[, 1]))),
                heatmap_legend_param = list(title=&quot;PP&quot;,at = c(0,0.5,1),
                                            lables = c(0,0.5,1)))
                
  
  pdf(out_name,width = width,height=height)
  draw(ht,heatmap_legend_side = &quot;right&quot;)
  dev.off()
  
  return(ht)
}</code></pre>
<pre class="r"><code>coloc_heatmap(&#39;output/coloc/coloc.ad.txt&#39;,threshold=0.7,height=6,width=7,out_name=&#39;output/figures/Figure4C.pdf&#39;)</code></pre>
</div>
<div id="figure-4d" class="section level2">
<h2>Figure 4D</h2>
<pre class="r"><code>coloc_heatmap(&#39;output/coloc/coloc.pd.txt&#39;,threshold=0.7,height=6,width=8.5,out_name=&#39;output/figures/Figure4D.pdf&#39;)</code></pre>
</div>
<div id="figure-4e" class="section level2">
<h2>Figure 4E</h2>
<pre class="r"><code>coloc_heatmap(&#39;output/coloc/coloc.ms.txt&#39;,threshold=0.7,height=14,width=7,out_name=&#39;output/figures/Figure4E.pdf&#39;)</code></pre>
</div>
<div id="figure-4f" class="section level2">
<h2>Figure 4F</h2>
<pre class="r"><code>coloc_heatmap(&#39;output/coloc/coloc.scz.txt&#39;,threshold=0.7,height=20,width=8,out_name=&#39;output/figures/Figure4F.pdf&#39;)</code></pre>
</div>
</div>
<div id="suplementary-figures" class="section level1">
<h1>Suplementary figures</h1>
<div id="figure-s1---mean-expression-markers" class="section level2">
<h2>Figure S1 - Mean expression markers</h2>
<pre class="r"><code>sum_expression &lt;- readRDS(&#39;data_sensitive/expression/cell_marker_expression.rds&#39;)</code></pre>
<pre class="r"><code>m &lt;- sum_expression %&gt;% 
  group_by(cell_type,symbol) %&gt;% 
  summarise(mean_cpm=mean(counts_scaled_1M,na.rm=TRUE)) %&gt;% 
  ungroup() %&gt;% 
  spread(cell_type,mean_cpm) %&gt;% 
  column_to_rownames(&#39;symbol&#39;) %&gt;% 
  as.matrix() </code></pre>
<pre class="r"><code>m &lt;- m[c(&#39;AQP4&#39;,&#39;CLDN5&#39;,&#39;SLC17A7&#39;,&#39;GAD2&#39;,&#39;C1QA&#39;,&#39;MOG&#39;,&#39;PDGFRA&#39;,&#39;RGS5&#39;),]</code></pre>
<pre class="r"><code>pheatmap::pheatmap(t(m),scale=&#39;row&#39;,cluster_cols = FALSE,cluster_rows = FALSE,angle_col = 0,filename = &#39;output/figures/FigureS1.png&#39;,height = 2,width=7)</code></pre>
<pre class="r"><code>pheatmap::pheatmap(t(m),scale=&#39;row&#39;,cluster_cols = FALSE,cluster_rows = FALSE,angle_col = 0)</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-83-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-s2---pi1-plots-all" class="section level2">
<h2>Figure S2 - Pi1 plots all</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;)

pi1 &lt;- d %&gt;% group_by(cell_type) %&gt;% summarise(pi1=1-qvalue::qvalue(bpval)$pi0) %&gt;% mutate(pi1_label=paste0(&#39;pi1 = &#39;,round(pi1,digits=2)))

p &lt;- d %&gt;% ggplot(.,aes(bpval,fill=cell_type)) + geom_histogram(bins=30, breaks = seq(0, 1, by=1/30)) + facet_wrap(~cell_type) + theme_classic() + theme(legend.position = &#39;none&#39;) + geom_text(data=pi1,aes(x=0.5,y=700,label=pi1_label))

ggsave(p,filename=&#39;output/figures/FigureS2.png&#39;,width=8,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-84-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-s3---metabrain" class="section level2">
<h2>Figure S3 - Metabrain</h2>
<div id="shared-eqtl-per-cell-type" class="section level3">
<h3>Shared eQTL per cell type</h3>
<pre class="r"><code>matched_snps &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.metabrain.matched.txt&#39;)</code></pre>
<pre class="r"><code>matched_snps_text_prop &lt;- matched_snps %&gt;% 
  add_count(cell_type) %&gt;%
  group_by(cell_type) %&gt;% 
  summarise(n_eqtl=unique(n),
            prop=paste0(&#39;prop = &#39;,round(sum(PValue&lt;0.05)*100/sum(PValue&lt;10),digits=1),&#39;%&#39;),
            pi1=paste0(&#39;pi1 = &#39;,ifelse(n_eqtl&lt;100,&#39;NA&#39;,
                                       round(1-qvalue::qvalue(PValue,)$pi0,digits=2))),
            label=paste0(prop,&#39;\n&#39;,pi1))</code></pre>
<pre class="r"><code>p &lt;- ggplot(matched_snps,aes(PValue,fill=cell_type)) + 
  geom_histogram() + 
  facet_wrap(~cell_type,scales=&#39;free_y&#39;,nrow=2) + 
  theme_classic() + 
  theme(legend.position = &#39;none&#39;) + 
  ggtitle(&#39;Metabrain Cortex matched pvalues&#39;) + 
  geom_text(data=matched_snps_text_prop,aes(x=0.5,y=Inf,vjust=1.5,label=label))

ggsave(p,filename=&#39;output/figures/FigureS3A.png&#39;,width=12,height=3,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-87-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="non-replicating-eqtl" class="section level3">
<h3>non-replicating eQTL</h3>
<pre class="r"><code>matched_snp_short &lt;- matched_snps %&gt;% 
  mutate(replication_status=
           factor(ifelse(PValue&lt;0.05,&#39;Replication p&lt;0.05&#39;,&#39;Replication p&gt;0.05&#39;))) %&gt;% 
  dplyr::select(ensembl,cell_type,SNP,replication_status)</code></pre>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;) %&gt;% 
  filter(adj_p&lt;0.05) %&gt;% 
  dplyr::rename(SNP=sid) %&gt;% 
  separate(pid,into=c(&#39;symbol&#39;,&#39;ensembl&#39;),sep=&#39;_&#39;)</code></pre>
<pre class="r"><code>d_sig &lt;- inner_join(d,matched_snp_short,by=c(&#39;ensembl&#39;,&#39;cell_type&#39;,&#39;SNP&#39;))</code></pre>
<pre class="r"><code>p &lt;- ggplot(d_sig,aes(replication_status,abs(dist+1),col=replication_status)) + 
  ggbeeswarm::geom_quasirandom(size=0.1) + 
  geom_boxplot(alpha=0.1,outlier.shape = NA,col=&#39;black&#39;,width=0.25) + 
  facet_wrap(~cell_type,nrow=2) + 
  scale_y_log10(expand=c(0.4,0)) +
  ggpubr::stat_compare_means(vjust = -0.5) +
  theme_classic() + 
  theme(legend.position=&#39;none&#39;)  + 
  ylab(&#39;Absolute distance to TSS (log10)&#39;) + 
  xlab(&#39;&#39;)

ggsave(p,filename=&#39;output/figures/FigureS3C.png&#39;,width=12,height=3,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-91-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="loeuf" class="section level3">
<h3>Loeuf</h3>
<pre class="r"><code>loeuf &lt;- read_tsv(&#39;data/gnomad_loeuf/supplementary_dataset_11_full_constraint_metrics.tsv&#39;) %&gt;% 
  filter(canonical==TRUE) %&gt;% 
  dplyr::select(gene,gene_id,transcript,oe_lof_upper,p) %&gt;% 
  mutate(oe_lof_upper_bin = ntile(oe_lof_upper, 10)) %&gt;% 
  dplyr::rename(ensembl=gene_id) %&gt;% 
  dplyr::select(-gene,-transcript)</code></pre>
<pre class="r"><code>d_sig &lt;- inner_join(d_sig,loeuf,by=&#39;ensembl&#39;)</code></pre>
<pre class="r"><code>p &lt;- ggplot(d_sig,aes(replication_status,oe_lof_upper,col=replication_status)) +
  ggbeeswarm::geom_quasirandom(size=0.1) + 
  geom_boxplot(alpha=0.1,outlier.shape = NA,col=&#39;black&#39;,width=0.25) + 
  facet_wrap(~cell_type,nrow=2) +
  scale_y_continuous(expand=c(0.4,0)) + 
  ggpubr::stat_compare_means(vjust = -0.5) +
  theme_classic() + 
  theme(legend.position=&#39;none&#39;)  + 
  ylab(&quot;LoF observed/expected upper bound \nfraction (LOEUF)&quot;) + 
  xlab(&#39;&#39;)

ggsave(p,filename=&#39;output/figures/FigureS3B.png&#39;,width=12,height=3,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-94-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s4s5-eqtl-location" class="section level2">
<h2>Figure S4/S5 eQTL location</h2>
<pre class="r"><code>snp_positions_file &lt;-  &#39;data_sensitive/genotypes/processed/snp_pos_hg38.txt&#39;
snp_coordinates &lt;- data.table::fread(snp_positions_file,header = FALSE,data.table = FALSE) %&gt;% 
  setNames(c(&#39;chr&#39;,&#39;start&#39;,&#39;sid&#39;)) %&gt;% 
  as_tibble()</code></pre>
<pre class="r"><code>dir.create(&#39;data_sensitive/eqtl/PC70/location&#39;,showWarnings = FALSE)
d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;) %&gt;%
 filter(adj_p&lt;0.05) %&gt;% 
 dplyr::select(cell_type,pid,sid) %&gt;% 
 unique() %&gt;% 
 left_join(.,snp_coordinates,by=&#39;sid&#39;) %&gt;% 
 mutate(start=as.numeric(start),
        end=start) %&gt;% 
 dplyr::select(chr,start,end,sid,pid,cell_type) %&gt;% 
 mutate(chr=paste0(&#39;chr&#39;,chr)) %&gt;% 
 arrange(chr,start,end) %&gt;% 
 write_tsv(.,&#39;data_sensitive/eqtl/PC70/location/eqtl_5FDR.bed&#39;,col_names = FALSE)</code></pre>
<pre class="r"><code>gtf &lt;- rtracklayer::import(&#39;data/gencode/Homo_sapiens.GRCh38.96.filtered.gtf&#39;) %&gt;% 
  as.data.frame() %&gt;% 
  filter(type==&#39;gene&#39;) %&gt;% 
  mutate(gene_label=gsub(&#39;\\..+&#39;,&#39;&#39;,gene_id)) %&gt;% 
  mutate(score=&#39;.&#39;) %&gt;% 
  dplyr::select(seqnames,start,end,gene_label,score,strand) %&gt;% 
  filter(seqnames %in% paste0(1:22)) %&gt;% 
  mutate(seqnames=as.character(paste0(&#39;chr&#39;,seqnames))) %&gt;% 
  arrange(seqnames,start,end) %&gt;% 
  write_tsv(&#39;data_sensitive/eqtl/PC70/location/Homo_sapiens.GRCh38.96.filtered.1_22.bed&#39;,col_names = FALSE)</code></pre>
<pre class="bash"><code>cd data_sensitive/eqtl/PC70/location/
ml bedtools/2.25.0-goolf-1.7.20
bedtools closest -D b -k 500 -a eqtl_5FDR.bed -b Homo_sapiens.GRCh38.96.filtered.1_22.bed &gt; eqtl_5FDR.closest_gene.bed</code></pre>
<pre class="r"><code>closest &lt;- data.table::fread(&#39;data_sensitive/eqtl/PC70/location/eqtl_5FDR.closest_gene.bed&#39;,header = FALSE,data.table = FALSE) %&gt;% 
  dplyr::select(-V7,-V8,-V9,-V11,-V12) %&gt;% 
  setNames(c(colnames(d),&#39;closest_gene&#39;,&#39;distance&#39;)) %&gt;% 
  group_by(sid,pid,cell_type) %&gt;% 
  mutate(rank=rank(abs(distance),ties.method=&#39;min&#39;)) %&gt;%
  ungroup() %&gt;% 
  separate(pid,into=c(&#39;symbol&#39;,&#39;ensembl&#39;),sep=&#39;_&#39;,remove = FALSE)</code></pre>
<div id="gene-body-location" class="section level3">
<h3>Gene body location</h3>
<pre class="r"><code>p &lt;- closest %&gt;% 
  filter(ensembl==closest_gene) %&gt;% 
  mutate(location=case_when(
  distance &lt; 0 ~ &#39;upstream&#39;,
  distance == 0 ~ &#39;gene body&#39;,
  distance &gt; 0 ~ &#39;downstream&#39;)) %&gt;% 
  dplyr::count(cell_type,location) %&gt;% 
  group_by(cell_type) %&gt;% 
  mutate(tot=sum(n),prop=round(n*100/tot,digits=2)) %&gt;% 
  ggplot(.,aes(reorder(cell_type,n),n,fill=location)) +
  geom_col(position=position_dodge()) + 
  coord_flip() + 
  guides(fill = guide_legend(reverse=T)) + 
  theme_classic() + 
  ylab(&#39;Number of cis-eQTL (5% FDR)&#39;) + 
  xlab(&#39;&#39;)
ggsave(p,filename=&#39;output/figures/FigureS4.png&#39;,width=5,height=3,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-100-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="rank-of-affected-gene" class="section level3">
<h3>Rank of affected gene</h3>
<pre class="r"><code>prop_closest &lt;- closest %&gt;% filter(ensembl==closest_gene) %&gt;% 
  mutate(rank=ifelse(rank&gt;5,&#39;&gt;5&#39;,rank)) %&gt;% 
  dplyr::count(cell_type,rank) %&gt;% 
  add_count(cell_type) %&gt;% 
  mutate(prop=round(n*100/nn,digits=2)) %&gt;% 
  mutate(rank=factor(rank,levels=c(&#39;1&#39;,&#39;2&#39;,&#39;3&#39;,&#39;4&#39;,&#39;5&#39;,&#39;&gt;5&#39;))) %&gt;% 
  arrange(cell_type,rank)</code></pre>
<pre class="r"><code>p &lt;- ggplot(prop_closest,aes(rank,prop,fill=rank)) + 
  geom_col() + 
  facet_wrap(~cell_type) + 
  theme_classic() + 
  theme(legend.position = &#39;none&#39;) + 
  xlab(&#39;Rank&#39;) + 
  ylab(&#39;Proportion of eQTL&#39;)
ggsave(p,filename=&#39;output/figures/FigureS5.png&#39;,width=6,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-102-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s6---eqtl-expression-level" class="section level2">
<h2>Figure S6 - eQTL expression level</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;)</code></pre>
<pre class="r"><code>exp_files &lt;- list.files(&#39;data_sensitive/eqtl/&#39;,pattern = &#39;.bed.gz$&#39;,full.names = T)</code></pre>
<pre class="r"><code>read_exp_get_median &lt;- function(i){
  
  cell_type &lt;- basename(exp_files[i]) %&gt;% 
    gsub(&#39;\\.bed.gz&#39;,&#39;&#39;,.) %&gt;% 
    gsub(&#39;\\.&#39;,&#39; &#39;,.) %&gt;% 
    gsub(&#39;OPCs   COPs&#39;,&#39;OPCs / COPs&#39;,.)
  
  exp_df &lt;- read_tsv(exp_files[i]) %&gt;% dplyr::select(-`#Chr`,-start,-end) %&gt;% column_to_rownames(&#39;ID&#39;) %&gt;% as.matrix()
  medians_per_gene &lt;- matrixStats::rowMedians(exp_df) %&gt;% 
    setNames(rownames(exp_df)) %&gt;% 
    as.data.frame() %&gt;% 
    setNames(&#39;median_cpm&#39;) %&gt;% 
    rownames_to_column(&#39;pid&#39;) %&gt;% 
    mutate(cell_type=cell_type)
  return(medians_per_gene)
}</code></pre>
<pre class="r"><code>median_expression_all &lt;- mclapply(1:length(exp_files),function(i) read_exp_get_median(i),mc.cores = 8) %&gt;% bind_rows()</code></pre>
<pre class="r"><code>d_cpm &lt;- left_join(d,median_expression_all,by=c(&#39;cell_type&#39;,&#39;pid&#39;)) %&gt;% mutate(Significant=ifelse(adj_p&lt;0.05,TRUE,FALSE))</code></pre>
<pre class="r"><code>p2 &lt;- ggplot(d_cpm,aes(cell_type,log2(median_cpm+1),col=Significant)) + 
  theme_classic() + 
  ggbeeswarm::geom_quasirandom(dodge.width = 0.9,size=0.5) + 
  ggpubr::stat_compare_means(aes(group = Significant),label = &quot;p.format&quot;) + 
  ylab(&#39;log2(median counts per million +1)&#39;) + 
  xlab(&#39;&#39;) +
  stat_summary(fun = median, 
                 fun.min = median, 
                 fun.max = median,
                 geom = &quot;crossbar&quot;,
                 width = 0.5,
                 size=0.25,
                 color=&#39;black&#39;,
                 position = position_dodge(width = 0.9),
                 aes(group=Significant)
  )</code></pre>
<pre class="r"><code>d_cpm &lt;- d_cpm %&gt;% mutate(type=ifelse(grepl(&#39;neurons&#39;,cell_type),&#39;Neurons&#39;,&#39;Glia&#39;)) %&gt;% 
  group_by(type,pid) %&gt;% 
  mutate(median_cpm2=median(median_cpm))</code></pre>
<pre class="r"><code>p1 &lt;- ggplot(d_cpm,aes(type,log2(median_cpm2+1),col=Significant)) + 
  theme_classic() + 
  ggbeeswarm::geom_quasirandom(size=0.5,dodge.width = 0.9) + 
  ggpubr::stat_compare_means(aes(group = Significant),label = &quot;p.format&quot;) + 
  ylab(&#39;log2(median counts per million +1)&#39;) + 
  xlab(&#39;&#39;) +
  stat_summary(fun = median, 
                 fun.min = median, 
                 fun.max = median,
                 geom = &quot;crossbar&quot;,
                 width = 0.5,
                 size=0.25,
                 color=&#39;black&#39;,
                 position = position_dodge(width = 0.9),
                 aes(group=Significant)) +
  theme(legend.position=&#39;none&#39;)</code></pre>
<pre class="r"><code>p &lt;- p1 + p2 +  plot_layout(widths = c(1, 4))
ggsave(p,filename=&#39;output/figures/FigureS6.png&#39;,width=14,height=5,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-111-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-s7---independent-eqtls" class="section level2">
<h2>Figure S7 - Independent eQTLs</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.70PCs.indep.txt&#39;)</code></pre>
<pre class="r"><code>p &lt;- d %&gt;% group_by(cell_type,phe_id) %&gt;% 
    summarise(max_rank=max(rank)+1) %&gt;%
    ungroup() %&gt;% 
  dplyr::count(cell_type,max_rank) %&gt;% 
  filter(max_rank==2) %&gt;% 
  ggplot(.,aes(reorder(cell_type,n),n,fill=reorder(cell_type,n)))+geom_col() +coord_flip() + theme_classic() + ylab(&#39;Number of genes with an independent cis-eQTL (5%FDR)&#39;) + xlab(&#39;&#39;) + theme(legend.position = &#39;none&#39;) + geom_text(aes(label=n),hjust=1.2)
ggsave(p,filename=&#39;output/figures/FigureS7B.png&#39;,width=6,height=2.5,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-113-1.png" width="672" style="display: block; margin: auto;" /></p>
<div id="distance-to-tss" class="section level3">
<h3>Distance to TSS</h3>
<pre class="r"><code>p &lt;- filter(d,rank==1) %&gt;% 
  ggplot(.,aes(dist_phe_var,fill=cell_type)) + 
  geom_histogram() + 
  facet_wrap(~cell_type) + 
  theme_classic() + 
  theme(legend.position=&#39;None&#39;) +
  xlab(&#39;Distance to TSS&#39;)
ggsave(p,filename=&#39;output/figures/FigureS7C.png&#39;,width=7,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-114-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>p &lt;- d %&gt;% 
  mutate(`n indep eQTL`=as.factor(rank+1)) %&gt;% 
  ggplot(.,aes(`n indep eQTL`,abs(dist_phe_var+1),group=`n indep eQTL`,col=`n indep eQTL`)) +
  ggbeeswarm::geom_quasirandom() + 
  geom_boxplot(alpha=0.1,width=0.1,col=&#39;black&#39;,outlier.shape = NA) + 
  theme_classic() + 
  xlab(&#39;eQTL rank&#39;) +
  ylab(&#39;Absolute distance to TSS&#39;) +
  theme(legend.position = &#39;none&#39;)  +
  scale_y_log10(expand = c(0.1, 0),breaks = trans_breaks(&quot;log10&quot;, function(x) 10^x),labels = trans_format(&quot;log10&quot;, math_format(10^.x))) + 
  ggpubr::stat_compare_means(label.x = 1.1,label.y = 6.3,comparisons = list(c(&#39;1&#39;,&#39;2&#39;)))

ggsave(p,filename=&#39;output/figures/FigureS7D.png&#39;,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-115-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s8---caveman" class="section level2">
<h2>Figure S8 - CaVEMan</h2>
<div id="epigenome-overlap" class="section level3">
<h3>Epigenome overlap</h3>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.70PCs.caveman.txt&#39;)</code></pre>
<pre class="r"><code>path &lt;-  system.file(package=&quot;liftOver&quot;, &quot;extdata&quot;, &quot;hg38ToHg19.over.chain&quot;)
ch &lt;-  import.chain(path)</code></pre>
<pre class="r"><code>dir.create(&#39;data_sensitive/eqtl/PC70_caveman/epigenome_overlap&#39;,showWarnings = FALSE)
bed_hg19 &lt;- d %&gt;% separate(GENE,into=c(&#39;symbol&#39;,&#39;ensembl&#39;,&#39;chr&#39;,&#39;start&#39;,&#39;A1&#39;,&#39;A2&#39;),sep=&#39;_&#39;) %&gt;% 
  mutate(chr=paste0(&#39;chr&#39;,chr)) %&gt;% 
  mutate(end=start) %&gt;% 
  dplyr::select(chr,start,end,cell_type,symbol,ensembl,Probability) %&gt;% 
  GenomicRanges::makeGRangesFromDataFrame(., TRUE) %&gt;% 
  liftOver(.,ch) %&gt;% 
  unlist() %&gt;% 
  as.data.frame() %&gt;% 
  as_tibble() %&gt;% 
  arrange(seqnames,start) %&gt;%
  unique() %&gt;% 
  write_tsv(.,&#39;data_sensitive/eqtl/PC70_caveman/epigenome_overlap/finemapped_SNPs.bed&#39;,col_names = FALSE)</code></pre>
<pre class="bash"><code>cd data_sensitive/eqtl/PC70_caveman/epigenome_overlap/
source ~/.bashrc
ml bedtools/2.25.0-goolf-1.7.20

ln -s ../../../../data/gwas_epigenome_overlap/corces_hg19.bed   .
bedtools intersect -loj -wa -wb -a finemapped_SNPs.bed -b corces_hg19.bed &gt; finemapped_epigenomic_marks.bed</code></pre>
<pre class="r"><code>d &lt;- read_tsv(&#39;data_sensitive/eqtl/PC70_caveman/epigenome_overlap/finemapped_epigenomic_marks.bed&#39;,col_names = FALSE) %&gt;% setNames(c(colnames(bed_hg19),&#39;chr_epi&#39;,&#39;start_epi&#39;,&#39;end_epi&#39;,&#39;study&#39;,&#39;cell_type_epi&#39;,&#39;specific&#39;,&#39;Annotation&#39;)) %&gt;% 
  mutate(id=paste(seqnames,start,end,cell_type,symbol,sep=&#39;_&#39;)) %&gt;% 
  dplyr::select(-width,-strand) %&gt;% 
  mutate(has_overlap=ifelse(chr_epi!=&#39;.&#39;,&#39;Overlap&#39;,&#39;No overlap&#39;))</code></pre>
<pre class="r"><code>p &lt;- d %&gt;% 
  mutate(any_overlap=ifelse(chr_epi==&#39;.&#39;,&#39;no_overlap&#39;,&#39;overlap&#39;)) %&gt;% 
  dplyr::select(id,any_overlap,Probability) %&gt;% unique() %&gt;% 
  group_by(any_overlap) %&gt;% 
  summarise(`All`=n(),
           `Prob &gt;0.25`=sum(Probability&gt;0.25),
           `Prob &gt;0.5`=sum(Probability&gt;0.5),
           `Prob &gt;0.75`=sum(Probability&gt;0.75),
           `Prob &gt;0.9`=sum(Probability&gt;0.9)) %&gt;% 
  gather(filter,value,-any_overlap) %&gt;% 
  group_by(filter) %&gt;% 
  mutate(tot=sum(value)) %&gt;% 
  mutate(prop=value/tot) %&gt;% 
  filter(any_overlap==&#39;overlap&#39;) %&gt;% 
  ggplot(.,aes(filter,prop,fill=filter))+geom_col(position = position_dodge()) + xlab(&#39;&#39;) + ylab(&#39;Proportion of top fine mapped cis-eQTL\noverlapping an epigenomic mark&#39;)  + theme_classic() + guides(fill=guide_legend(reverse = TRUE)) + scale_fill_hue(direction=-1) + theme(legend.position=&#39;none&#39;,text=element_text(size=20)) + coord_flip()

ggsave(p,filename=paste0(&#39;output/figures/FigureS8A.png&#39;),width=7,height=4,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-121-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="histogram" class="section level3">
<h3>Histogram</h3>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.70PCs.caveman.txt&#39;)</code></pre>
<pre class="r"><code>p &lt;- ggplot(d,aes(Probability,fill=cell_type)) + 
  geom_histogram() + 
  theme_classic() + 
  xlab(&#39;CaVEMaN Causal probability&#39;) + 
  theme(legend.position=&#39;none&#39;) + 
  facet_wrap(~cell_type,scales=&#39;free_y&#39;)
ggsave(p,filename=&#39;output/figures/FigureS8B.png&#39;,width=5,height=3,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-123-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s9---epigenome-enrichment" class="section level2">
<h2>Figure S9 - Epigenome enrichment</h2>
<div id="nott" class="section level3">
<h3>Nott</h3>
<pre class="r"><code>fdensity_files &lt;- list.files(&#39;data/epigenome_enrichment/fdensity/results/nott/&#39;,full.names = TRUE)</code></pre>
<pre class="r"><code>fdensity_res &lt;- tibble(fdensity_files) %&gt;% mutate(file_content=map(fdensity_files,read_delim,delim=&#39; &#39;,col_names=FALSE)) %&gt;% 
  unnest(file_content) %&gt;% 
  mutate(name=basename(fdensity_files)) %&gt;% 
  dplyr::select(-fdensity_files) %&gt;% 
  mutate(name=gsub(&#39;_specific&#39;,&#39;.specific&#39;,name)) %&gt;% 
  separate(name,into=c(&#39;tmp&#39;,&#39;sig&#39;,&#39;annot&#39;),sep=&#39;_&#39;) %&gt;% 
  dplyr::select(-tmp) %&gt;% 
  mutate(sig=gsub(&#39;.sig5FDR.bed&#39;,&#39;&#39;,sig)) %&gt;% 
  mutate(annot=gsub(&#39;.bed&#39;,&#39;&#39;,annot)) %&gt;% 
  group_by(sig,annot) %&gt;% 
  mutate(X3_norm=X3/sum(X3)) %&gt;% 
  mutate(sig=gsub(&#39;\\.&#39;,&#39; &#39;,sig) %&gt;% gsub(&#39;OPCs   COPs&#39;,&#39;OPCs / COPs&#39;,.)) %&gt;% 
  mutate(sig=paste0(sig,&#39; cis-eQTL&#39;)) %&gt;% 
  mutate(sig=factor(sig,levels=c(&#39;Astrocytes cis-eQTL&#39;,&#39;Microglia cis-eQTL&#39;,&#39;Oligodendrocytes cis-eQTL&#39;,&#39;OPCs / COPs cis-eQTL&#39;,&#39;Excitatory neurons cis-eQTL&#39;,&#39;Inhibitory neurons cis-eQTL&#39;,&#39;Pericytes cis-eQTL&#39;,&#39;Endothelial cells cis-eQTL&#39;))) %&gt;% 
  mutate(annot = gsub(&#39;.enhancer.specific&#39;, &#39; Enhancer (specific)&#39;,annot))</code></pre>
<pre class="r"><code>p &lt;- fdensity_res %&gt;% 
  filter(grepl(&#39;Enhancer&#39;,annot)) %&gt;% 
  filter(!sig%in%c(&#39;Endothelial cells cis-eQTL&#39;,&#39;Pericytes cis-eQTL&#39;)) %&gt;% 
  ungroup() %&gt;% 
  mutate(annot=factor(annot,
                      levels=c(&#39;Astrocytes Enhancer (specific)&#39;,
                               &#39;Microglia Enhancer (specific)&#39;,
                               &#39;Oligodendrocytes Enhancer (specific)&#39;,
                               &#39;Neuronal Enhancer (specific)&#39;))) %&gt;% 
  ggplot(.,aes((X1+X2)/2,X3_norm,col=sig)) + geom_line() + facet_wrap(sig~annot,ncol=4) + theme_classic() + theme(legend.position = &#39;none&#39;,axis.text.x = element_text(size=7),axis.text.y = element_text(size=7)) + xlab(&#39;Distance to cis-eQTL&#39;) + ylab(&#39;Density&#39;) + scale_x_continuous(breaks=c(-1000000,-500000,0,500000,1000000),labels = c(&#39;-1Mb&#39;,&#39;-500kb&#39;,&#39;0&#39;,&#39;500kb&#39;,&#39;1Mb&#39;))

ggsave(p,filename = &#39;output/figures/FigureS9.png&#39;,width=12,height=8,limitsize = FALSE,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-126-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s10---epigenome-enrichment" class="section level2">
<h2>Figure S10 - Epigenome enrichment</h2>
<div id="fullard" class="section level3">
<h3>Fullard</h3>
<pre class="r"><code>fdensity_files &lt;- list.files(&#39;data/epigenome_enrichment/fdensity/results/fullard/&#39;,full.names = TRUE)</code></pre>
<pre class="r"><code>fdensity_res &lt;- tibble(fdensity_files) %&gt;% 
  mutate(file_content=map(fdensity_files,read_delim,delim=&#39; &#39;,col_names=FALSE)) %&gt;% unnest(file_content) %&gt;% 
  mutate(name=basename(fdensity_files)) %&gt;% 
  dplyr::select(-fdensity_files) %&gt;% 
  mutate(name=gsub(&#39;__&#39;,&#39;_&#39;,name)) %&gt;% 
  separate(name,into=c(&#39;tmp&#39;,&#39;sig&#39;,&#39;region&#39;,&#39;annot&#39;),sep=&#39;_&#39;) %&gt;% 
  dplyr::select(-tmp) %&gt;% 
  mutate(sig=gsub(&#39;.sig5FDR.bed&#39;,&#39;&#39;,sig)) %&gt;% 
  mutate(annot=gsub(&#39;.bed&#39;,&#39;&#39;,annot)) %&gt;% 
  mutate(annot=paste0(region,&#39;_&#39;,annot)) %&gt;% 
  dplyr::select(-region) %&gt;% 
  group_by(sig,annot) %&gt;% 
  mutate(X3_norm=X3/sum(X3)) %&gt;% 
  mutate(sig=gsub(&#39;\\.&#39;,&#39; &#39;,sig) %&gt;% gsub(&#39;OPCs   COPs&#39;,&#39;OPCs / COPs&#39;,.)) %&gt;% 
  mutate(sig=paste0(sig,&#39; cis-eQTL&#39;)) %&gt;% 
  mutate(sig=factor(sig,levels=c(&#39;Astrocytes cis-eQTL&#39;,&#39;Microglia cis-eQTL&#39;,&#39;Oligodendrocytes cis-eQTL&#39;,&#39;OPCs / COPs cis-eQTL&#39;,&#39;Excitatory neurons cis-eQTL&#39;,&#39;Inhibitory neurons cis-eQTL&#39;,&#39;Pericytes cis-eQTL&#39;,&#39;Endothelial cells cis-eQTL&#39;)))</code></pre>
<pre class="r"><code>p &lt;- fdensity_res %&gt;% 
  filter(grepl(&#39;DLPFC&#39;,annot)) %&gt;% 
  filter(!sig%in%c(&#39;Pericytes cis-eQTL&#39;,&#39;Endothelial cells cis-eQTL&#39;)) %&gt;% 
  mutate(annot=gsub(&#39;DLPFC_|.specific&#39;,&#39;&#39;,annot)) %&gt;% 
  mutate(annot=paste0(Hmisc::capitalize(annot),&#39; ATAC (specific)&#39;)) %&gt;% 
  mutate(annot=factor(annot,levels=c(&#39;Glia ATAC (specific)&#39;,&#39;Neuron ATAC (specific)&#39;))) %&gt;% 
  ggplot(.,aes((X1+X2)/2,X3_norm,col=sig)) + geom_line() + facet_wrap(sig~annot,ncol=4) + theme_classic() + theme(legend.position = &#39;none&#39;,axis.text.x = element_text(size=7),axis.text.y = element_text(size=7)) + xlab(&#39;Distance to cis-eQTL&#39;) + ylab(&#39;Density&#39;) + scale_x_continuous(breaks=c(-1000000,-500000,0,500000,1000000),labels = c(&#39;-1Mb&#39;,&#39;-500kb&#39;,&#39;0&#39;,&#39;500kb&#39;,&#39;1Mb&#39;))

ggsave(p,filename = &#39;output/figures/FigureS10.png&#39;,width=8,height=6,limitsize = FALSE,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-129-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s11---cell-type-specific-eqtl" class="section level2">
<h2>Figure S11 - cell type specific eQTL</h2>
<div id="distance-to-tss-1" class="section level3">
<h3>Distance to TSS</h3>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl_specific/eqtl.PC70.specific.txt&#39;) %&gt;% 
  #Sets pvalue to NA if the model did not converge
  mutate(nb_pvalue_aggregate=ifelse(nb_pvalue_aggregate_model_converged==FALSE,NA,nb_pvalue_aggregate)) %&gt;% 
  mutate(nb_pvalue_at_least_one=ifelse(nb_pvalue_at_least_one_model_converged==FALSE,NA,nb_pvalue_at_least_one)) %&gt;%   #Remove genes for which the model did not converge (9 genes)
  filter(!is.na(nb_pvalue_aggregate),
         !is.na(nb_pvalue_at_least_one),
         nb_pvalue_at_least_one!=Inf) %&gt;% 
  #Get adjusted pvalues
  mutate(nb_pvalue_aggregate_adj=p.adjust(nb_pvalue_aggregate,method=&#39;fdr&#39;),
         nb_pvalue_at_least_one_adj=p.adjust(nb_pvalue_at_least_one,method = &#39;fdr&#39;)) %&gt;% 
  #For each row, get the maximum pvalue across all cell types, this will be the gene-level pvalue testing whether the genetic effect on gene expression is different than all other cell types
  rowwise() %&gt;% 
  mutate(nb_pvalue_sig_all=max(Astrocytes_p,`Endothelial cells_p`,`Excitatory neurons_p`,`Inhibitory neurons_p`,Microglia_p,Oligodendrocytes_p,`OPCs / COPs_p`,Pericytes_p,na.rm=TRUE)) %&gt;% 
  ungroup() %&gt;% 
  mutate(nb_pvalue_all_adj = p.adjust(nb_pvalue_sig_all,method=&#39;fdr&#39;))</code></pre>
<pre class="r"><code>d_short &lt;- d %&gt;% 
  dplyr::select(cell_type_id,gene_id,snp_id,nb_pvalue_aggregate_adj,nb_pvalue_at_least_one_adj,nb_pvalue_all_adj) %&gt;% 
  gather(test,p_adj,nb_pvalue_aggregate_adj:nb_pvalue_all_adj) %&gt;% 
  mutate(specific=ifelse(p_adj&lt;0.05,&#39;Specific&#39;,&#39;Shared&#39;)) %&gt;% 
  mutate(test=factor(case_when(
    test==&#39;nb_pvalue_aggregate_adj&#39; ~ &#39;Aggregate&#39;,
    test==&#39;nb_pvalue_at_least_one_adj&#39; ~ &#39;At least one&#39;,
    test==&#39;nb_pvalue_all_adj&#39; ~ &#39;All&#39;),levels=c(&#39;Aggregate&#39;,&#39;At least one&#39;,&#39;All&#39;)))</code></pre>
<pre class="r"><code>eqtl &lt;- read_tsv(&#39;output/eqtl/eqtl.PC70.txt&#39;) %&gt;% 
  filter(adj_p&lt;0.05) %&gt;% 
  dplyr::select(cell_type,pid,dist)</code></pre>
<pre class="r"><code>d_short &lt;- inner_join(d_short,eqtl,by=c(&#39;gene_id&#39;=&#39;pid&#39;,&#39;cell_type_id&#39;=&#39;cell_type&#39;)) %&gt;% 
  mutate(dist=ifelse(dist==0,1,dist))</code></pre>
<pre class="r"><code>p &lt;- ggplot(d_short,aes(specific,abs(dist),col=specific)) + geom_boxplot(width=0.2,outlier.shape = NA) + ggbeeswarm::geom_quasirandom(size=0.1,alpha=0.3) + theme_classic() + xlab(&#39;&#39;) + ylab(&quot;Log10(absolute distance to TSS)&quot;) + scale_y_log10(expand=c(0.1,0)) + theme(text=element_text(size=20)) + ggpubr::stat_compare_means(vjust=-1) + theme(legend.position = &#39;none&#39;) + facet_wrap(~test)
ggsave(p,filename=&#39;output/figures/FigureS11.png&#39;,width=12,height=5,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-134-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s12---cell-type-specific-eqtl" class="section level2">
<h2>Figure S12 - cell type specific eQTL</h2>
<div id="ploeuf" class="section level3">
<h3>PLOEUF</h3>
<pre class="r"><code>loeuf &lt;- read_tsv(&#39;data/gnomad_loeuf/supplementary_dataset_11_full_constraint_metrics.tsv&#39;) %&gt;% 
  filter(canonical==TRUE) %&gt;% 
  dplyr::select(gene,gene_id,transcript,oe_lof_upper,p) %&gt;% 
  mutate(oe_lof_upper_bin = ntile(oe_lof_upper, 10)) %&gt;% 
  dplyr::rename(ensembl=gene_id) %&gt;% 
  dplyr::select(-gene,-transcript)</code></pre>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl_specific/eqtl.PC70.specific.txt&#39;) %&gt;% 
  #Sets pvalue to NA if the model did not converge
  mutate(nb_pvalue_aggregate=ifelse(nb_pvalue_aggregate_model_converged==FALSE,NA,nb_pvalue_aggregate)) %&gt;% 
  mutate(nb_pvalue_at_least_one=ifelse(nb_pvalue_at_least_one_model_converged==FALSE,NA,nb_pvalue_at_least_one)) %&gt;%   #Remove genes for which the model did not converge (9 genes)
  filter(!is.na(nb_pvalue_aggregate),
         !is.na(nb_pvalue_at_least_one),
         nb_pvalue_at_least_one!=Inf) %&gt;% 
  #Get adjusted pvalues
  mutate(nb_pvalue_aggregate_adj=p.adjust(nb_pvalue_aggregate,method=&#39;fdr&#39;),
         nb_pvalue_at_least_one_adj=p.adjust(nb_pvalue_at_least_one,method = &#39;fdr&#39;)) %&gt;% 
  #For each row, get the maximum pvalue across all cell types, this will be the gene-level pvalue testing whether the genetic effect on gene expression is different than all other cell types
  rowwise() %&gt;% 
  mutate(nb_pvalue_sig_all=max(Astrocytes_p,`Endothelial cells_p`,`Excitatory neurons_p`,`Inhibitory neurons_p`,Microglia_p,Oligodendrocytes_p,`OPCs / COPs_p`,Pericytes_p,na.rm=TRUE)) %&gt;% 
  ungroup() %&gt;% 
  mutate(nb_pvalue_all_adj = p.adjust(nb_pvalue_sig_all,method=&#39;fdr&#39;))</code></pre>
<pre class="r"><code>d_short &lt;- d %&gt;% 
  dplyr::select(cell_type_id,gene_id,snp_id,nb_pvalue_aggregate_adj,nb_pvalue_at_least_one_adj,nb_pvalue_all_adj) %&gt;% 
  gather(test,p_adj,nb_pvalue_aggregate_adj:nb_pvalue_all_adj) %&gt;% 
  mutate(specific=ifelse(p_adj&lt;0.05,&#39;Specific&#39;,&#39;Shared&#39;)) %&gt;% 
  mutate(test=factor(case_when(
    test==&#39;nb_pvalue_aggregate_adj&#39; ~ &#39;Aggregate&#39;,
    test==&#39;nb_pvalue_at_least_one_adj&#39; ~ &#39;At least one&#39;,
    test==&#39;nb_pvalue_all_adj&#39; ~ &#39;All&#39;),levels=c(&#39;Aggregate&#39;,&#39;At least one&#39;,&#39;All&#39;))) %&gt;% 
  separate(gene_id,into=c(&#39;symbol&#39;,&#39;ensembl&#39;),sep=&#39;_&#39;)</code></pre>
<pre class="r"><code>d_short &lt;- d_short %&gt;% left_join(.,loeuf,by=&#39;ensembl&#39;)</code></pre>
<pre class="r"><code>plot_loeuf_int &lt;- function(test_var){
  
  #Get only results from test variable (aggregate, at least one or all)
  d_short_tmp &lt;- d_short %&gt;% 
    filter(test==test_var)
  
  #Duplicate the data and set cell_type ids to All (all cell types)
  d_short_all &lt;- d_short_tmp %&gt;% 
    mutate(cell_type_id=&#39;All&#39;)
  
  d_short_plot &lt;- rbind(d_short_tmp,d_short_all)
  
  p &lt;- d_short_plot %&gt;% 
    ggplot(.,aes(specific,oe_lof_upper,col=specific)) + 
    ggbeeswarm::geom_quasirandom(size=0.1) + 
    ggpubr::stat_compare_means(vjust = -1) + 
    facet_wrap(~cell_type_id,nrow=1) + 
    theme_classic()+ 
    stat_summary(fun = median, 
                 fun.min = median, 
                 fun.max = median,
                 geom = &quot;crossbar&quot;,
                 width = 0.5,
                 size=0.25,
                 color=&#39;black&#39;) + 
    ylab(&quot;LoF observed/expected upper bound \nfraction (LOEUF)&quot;) +
    theme(legend.position=&#39;none&#39;)+
    scale_y_continuous(expand=c(0.25,0)) +
    xlab(&#39;&#39;) +
    ggtitle(test_var)
    
}</code></pre>
<pre class="r"><code>p &lt;- plot_loeuf_int(&#39;Aggregate&#39;)
ggsave(p,filename=&#39;output/figures/FigureS12A.png&#39;,width=17,height=3,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-140-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>p &lt;- plot_loeuf_int(&#39;At least one&#39;)
ggsave(p,filename=&#39;output/figures/FigureS12B.png&#39;,width=17,height=3,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-141-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>p &lt;- plot_loeuf_int(&#39;All&#39;)
ggsave(p,filename=&#39;output/figures/FigureS12C.png&#39;,width=17,height=3,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-142-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s13---cell-type-specific-eqtl" class="section level2">
<h2>Figure S13 - cell type specific eQTL</h2>
<div id="magma" class="section level3">
<h3>MAGMA</h3>
<pre class="r"><code>entrez_2_ensembl &lt;- AnnotationDbi::toTable(org.Hs.eg.db::org.Hs.egENSEMBL) %&gt;% 
  add_count(gene_id,name = &#39;n_entrez&#39; ) %&gt;% add_count(ensembl_id,name=&#39;n_ensembl&#39;) %&gt;% 
  filter(n_entrez==1,n_ensembl==1) %&gt;% 
  dplyr::select(-n_entrez,-n_ensembl) %&gt;% 
  dplyr::rename(ensembl=ensembl_id,GENE=gene_id) %&gt;% 
  as_tibble()</code></pre>
<pre class="r"><code>dir.create(&#39;data/magma_interaction_eqtl/&#39;,showWarnings = FALSE)
d &lt;- read_tsv(&#39;output/eqtl_specific/eqtl.PC70.specific.txt&#39;) %&gt;% 
  #Sets pvalue to NA if the model did not converge
  mutate(nb_pvalue_aggregate=ifelse(nb_pvalue_aggregate_model_converged==FALSE,NA,nb_pvalue_aggregate)) %&gt;% 
  mutate(nb_pvalue_at_least_one=ifelse(nb_pvalue_at_least_one_model_converged==FALSE,NA,nb_pvalue_at_least_one)) %&gt;%   #Remove genes for which the model did not converge (9 genes)
  filter(!is.na(nb_pvalue_aggregate),
         !is.na(nb_pvalue_at_least_one),
         nb_pvalue_at_least_one!=Inf) %&gt;% 
  #Get adjusted pvalues
  mutate(nb_pvalue_aggregate_adj=p.adjust(nb_pvalue_aggregate,method=&#39;fdr&#39;),
         nb_pvalue_at_least_one_adj=p.adjust(nb_pvalue_at_least_one,method = &#39;fdr&#39;)) %&gt;% 
  #For each row, get the maximum pvalue across all cell types, this will be the gene-level pvalue testing whether the genetic effect on gene expression is different than all other cell types
  rowwise() %&gt;% 
  mutate(nb_pvalue_sig_all=
           max(Astrocytes_p,`Endothelial cells_p`,`Excitatory neurons_p`,`Inhibitory neurons_p`,Microglia_p,Oligodendrocytes_p,`OPCs / COPs_p`,Pericytes_p,na.rm=TRUE)) %&gt;% 
  ungroup() %&gt;% 
  mutate(nb_pvalue_all_adj = p.adjust(nb_pvalue_sig_all,method=&#39;fdr&#39;)) %&gt;% 
  separate(gene_id,into=c(&#39;symbol&#39;,&#39;ensembl&#39;),sep=&#39;_&#39;) %&gt;% 
  filter(nb_pvalue_all_adj&lt;0.05) %&gt;% 
  inner_join(.,entrez_2_ensembl,by=&#39;ensembl&#39;) %&gt;% 
  mutate(cell_type_id=make.names(cell_type_id)) %&gt;% 
  dplyr::select(GENE,cell_type_id) %&gt;% 
  write_tsv(.,path = &#39;data/magma_interaction_eqtl/cell_type_specific.txt&#39;,col_names =FALSE)</code></pre>
<pre class="bash"><code>cd data/magma_interaction_eqtl
source ~/.bashrc

##Gene sets
cell_type_specific_eqtl=&#39;cell_type_specific.txt&#39;

##GWAS
ms=&#39;../gwas/ms/ms.annotated_35kbup_10_down.genes.raw&#39;
ad=&#39;../gwas/ad/ad.annotated_35kbup_10_down.genes.raw&#39;
pd=&#39;../gwas/pd/parkinsonMeta2020.annotated_35kbup_10_down.genes.raw&#39;
scz=&#39;../gwas/scz/scz3.annotated_35kbup_10_down.genes.raw&#39;

##Run MAGMA
magma --gene-results $ms --set-annot $cell_type_specific_eqtl col=1,2  --out ms
magma --gene-results $ad --set-annot $cell_type_specific_eqtl col=1,2  --out ad
magma --gene-results $pd --set-annot $cell_type_specific_eqtl col=1,2  --out pd
magma --gene-results $scz --set-annot $cell_type_specific_eqtl col=1,2  --out scz</code></pre>
<pre class="r"><code>files &lt;- list.files(&#39;data/magma_interaction_eqtl&#39;,pattern=&#39;.gsa.out&#39;,full.names = T)</code></pre>
<pre class="r"><code>d &lt;- tibble(file_names=files) %&gt;% 
  mutate(trait=gsub(&#39;\\..+&#39;,&#39;&#39;,basename(file_names))) %&gt;% 
  mutate(file_content=map(file_names,read.table,header=T)) %&gt;%
  unnest(file_content) %&gt;% 
  group_by(trait) %&gt;% 
  mutate(fdr.local=p.adjust(P,method=&#39;fdr&#39;)) %&gt;% 
  mutate(Significance=ifelse(fdr.local&lt;=0.05,&#39;FDR&lt;5%&#39;,&#39;Not significant&#39;)) %&gt;% 
  mutate(trait=factor(case_when(
    trait==&#39;ad&#39; ~ &#39;Alzheimer&#39;,
    trait==&#39;ms&#39; ~ &#39;Multiple Sclerosis&#39;,
    trait==&#39;pd&#39; ~ &#39;Parkinson&#39;,
    trait==&#39;scz&#39; ~ &#39;Schizophrenia&#39;),levels=c(&#39;Alzheimer&#39;,&#39;Multiple Sclerosis&#39;,&#39;Parkinson&#39;,&#39;Schizophrenia&#39;))) %&gt;% 
  mutate(VARIABLE=gsub(&#39;\\.&#39;,&#39; &#39;,VARIABLE) %&gt;% gsub(&#39;OPCs   COPs&#39;,&#39;OPCs / COPs&#39;,.))</code></pre>
<pre class="r"><code>cbPalette &lt;- c(&quot;#D55E00&quot;,&quot;#56B4E9&quot;,&quot;#CC79A7&quot;)
#
p &lt;- ggplot(d,aes(VARIABLE,-log10(P),fill=Significance)) + geom_col() + facet_wrap(~trait) + coord_flip() + theme_classic() + xlab(&#39;&#39;)
ggsave(p,filename=&#39;output/figures/FigureS13.png&#39;,width=8,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-148-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s14---cell-type-specific-eqtl" class="section level2">
<h2>Figure S14 - cell type specific eQTL</h2>
<div id="examples" class="section level3">
<h3>Examples</h3>
<pre class="r"><code>hdF5_file_path &lt;- &#39;output/shiny/data.h5&#39;</code></pre>
<pre class="r"><code>eqtl_specific &lt;- h5read(hdF5_file_path, &quot;eqtl_results/eqtl_results_specific&quot;)</code></pre>
<pre class="r"><code>plot_eqtl_specific &lt;- function(cell_type_name,gene_name,snp_name){
  
  gene_short &lt;- gsub(&#39;_.+&#39;,&#39;&#39;,gene_name)
  
  genotype &lt;- h5read(hdF5_file_path, paste0(&quot;genotype/&quot;,snp_name))
  expression &lt;- h5read(hdF5_file_path, paste0(&quot;expression/&quot;,gene_name)) %&gt;% 
    dplyr::rename(individual=individual_id)
  
  pval &lt;- filter(eqtl_specific,gene==gene_name,cell_type==cell_type_name) %&gt;% pull(nb_pvalue_all_adj)
  
  d &lt;- inner_join(genotype,expression,by=&#39;individual&#39;) %&gt;% 
  mutate(genotype_label = case_when(
    genotype == 0 ~ paste0(REF,REF),
    genotype == 1 ~ paste0(REF,ALT),
    genotype == 2 ~ paste0(ALT,ALT)
  )) %&gt;%  mutate(cell_type=fct_relevel(cell_type,cell_type_name))
  
  p &lt;- ggplot(d, aes(genotype_label,log2_cpm,col=cell_type)) + 
    ggbeeswarm::geom_quasirandom(size=0.5) + 
      geom_boxplot(alpha=0.05,aes(group=genotype_label),outlier.shape = NA) + 
    theme_classic() + 
    theme(legend.position=&#39;none&#39;,text=element_text(size=16), plot.title = element_text(size=10)) + 
    xlab(&#39;Genotype&#39;) + 
    ylab(&#39;Expression (cpm) (log2+1)&#39;) +
    ggtitle(paste0(gene_short,&#39; - &#39;,snp_name,&#39; - &#39;,cell_type_name,&#39;\nadjusted pvalue = &#39;, pval)) + facet_wrap(~cell_type,ncol=4) + scale_color_discrete(limits=unique(d$cell_type) %&gt;% as.character() %&gt;% sort())
  p
}</code></pre>
<pre class="r"><code>gene_name &lt;- &#39;AVEN_ENSG00000169857&#39;
snp_name &lt;- &#39;rs8038713&#39;
cell_type_name &lt;- &#39;Oligodendrocytes&#39;
p &lt;- plot_eqtl_specific(cell_type_name,gene_name,snp_name) + theme(text=element_text(size=20),plot.title = element_text(size=12))
p + theme(text=element_text(size=14))</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-152-1.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(p,filename = &#39;output/figures/FigureS14A.png&#39;,width=9,height=5)</code></pre>
<pre class="r"><code>gene_name &lt;- &#39;ACAP2_ENSG00000114331&#39;
snp_name &lt;- &#39;rs6437374&#39;
cell_type_name &lt;- &#39;Astrocytes&#39;
p &lt;- plot_eqtl_specific(cell_type_name,gene_name,snp_name) + theme(text=element_text(size=20),plot.title = element_text(size=12))
p + theme(text=element_text(size=14))</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-153-1.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(p,filename = &#39;output/figures/FigureS14B.png&#39;,width=9,height=5)</code></pre>
</div>
</div>
<div id="figure-s15---cell-type-specific-eqtl" class="section level2">
<h2>Figure S15 - cell type specific eQTL</h2>
<div id="pi1" class="section level3">
<h3>Pi1</h3>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl_specific/eqtl.PC70.specific.txt&#39;) %&gt;% 
  #Sets pvalue to NA if the model did not converge
  mutate(nb_pvalue_aggregate=ifelse(nb_pvalue_aggregate_model_converged==FALSE,NA,nb_pvalue_aggregate)) %&gt;% 
  mutate(nb_pvalue_at_least_one=ifelse(nb_pvalue_at_least_one_model_converged==FALSE,NA,nb_pvalue_at_least_one)) %&gt;%   #Remove genes for which the model did not converge (9 genes)
  filter(!is.na(nb_pvalue_aggregate),
         !is.na(nb_pvalue_at_least_one),
         nb_pvalue_at_least_one!=Inf)</code></pre>
<pre class="r"><code>d_gather &lt;- d %&gt;% 
  dplyr::select(-snp_id) %&gt;% 
  gather(cell_type_test,p,-cell_type_id,-gene_id,-nb_pvalue_aggregate,-nb_pvalue_at_least_one,-nb_pvalue_aggregate_model_converged,-nb_pvalue_at_least_one_model_converged) %&gt;% 
  mutate(cell_type_test=gsub(&#39;_p&#39;,&#39;&#39;,cell_type_test)) %&gt;% 
  mutate(cell_type_id=factor(cell_type_id,levels=c(&#39;Microglia&#39;,&#39;OPCs / COPs&#39;, &#39;Astrocytes&#39;,&#39;Oligodendrocytes&#39;,&#39;Inhibitory neurons&#39;, &#39;Excitatory neurons&#39;,&#39;Endothelial cells&#39;,&#39;Pericytes&#39;)),
         cell_type_test=factor(cell_type_test,levels=c(&#39;Microglia&#39;,&#39;OPCs / COPs&#39;, &#39;Astrocytes&#39;,&#39;Oligodendrocytes&#39;,&#39;Inhibitory neurons&#39;, &#39;Excitatory neurons&#39;,&#39;Endothelial cells&#39;,&#39;Pericytes&#39;)))</code></pre>
<pre class="r"><code>pi1 &lt;- d_gather %&gt;% 
  filter(!is.na(p)) %&gt;% 
  filter(!cell_type_id%in%c(&#39;Endothelial cells&#39;,&#39;Pericytes&#39;)) %&gt;% 
  group_by(cell_type_id,cell_type_test) %&gt;% 
  summarise(pi1=1-qvalue(p)$pi0)</code></pre>
<pre class="r"><code>p &lt;- ggplot(d_gather,aes(p,fill=cell_type_id)) + geom_histogram() + facet_grid(cell_type_id~cell_type_test,scales = &#39;free_y&#39;) +theme_classic() + xlab(&#39;Interaction p-value&#39;) + theme(legend.position=&#39;none&#39;) + geom_text(data=pi1,aes(x=0.5,y=Inf,label=paste0(&#39;pi1=&#39;,round(pi1,digits=2)),vjust=1.5))
ggsave(p,filename=&#39;output/figures/FigureS15.png&#39;,width=14,height=10,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-157-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-s16---number-of-coloc-loci" class="section level2">
<h2>Figure S16 - Number of coloc loci</h2>
<pre class="r"><code>coloc_ms &lt;- read_tsv(&#39;output/coloc/coloc.ms.txt&#39;) %&gt;% mutate(trait=&#39;Multiple Sclerosis&#39;) 
coloc_ad &lt;- read_tsv(&#39;output/coloc/coloc.ad.txt&#39;) %&gt;% mutate(trait=&#39;Alzheimer&#39;)
coloc_scz &lt;- read_tsv(&#39;output/coloc/coloc.scz.txt&#39;) %&gt;% mutate(trait=&#39;Schizophrenia&#39;)
coloc_pd &lt;- read_tsv(&#39;output/coloc/coloc.pd.txt&#39;) %&gt;% mutate(trait=&#39;Parkinson&#39;)</code></pre>
<pre class="r"><code>coloc &lt;- rbind(coloc_ms,coloc_ad,coloc_scz,coloc_pd)</code></pre>
<pre class="r"><code>get_loci_explained &lt;- function(thresh){
  filter(coloc,PP.H4.abf &gt;= thresh) %&gt;% 
    group_by(trait) %&gt;% 
    summarise(n_loci_sig=length(unique(locus))) %&gt;% 
    mutate(threshold=thresh)
}</code></pre>
<pre class="r"><code>thresholds &lt;- c(0,0.7)
n_loci &lt;- lapply(thresholds,get_loci_explained) %&gt;% bind_rows() %&gt;% 
  mutate(threshold=factor(case_when(
    threshold == 0 ~ &#39;All loci&#39;,
    threshold == 0.7 ~ &#39;PP &gt;0.7&#39;,
  ),levels=c(&#39;All loci&#39;,&#39;PP &gt;0.7&#39;)))</code></pre>
<pre class="r"><code>p &lt;- n_loci %&gt;% filter(threshold%in%c(&#39;All loci&#39;,&#39;PP &gt;0.7&#39;)) %&gt;% 
  ggplot(.,aes(trait,n_loci_sig,fill=threshold)) + geom_col(position = position_dodge()) + facet_wrap(~trait,scales = &#39;free&#39;) + theme_classic() + theme(axis.ticks.x = element_blank(),axis.text.x = element_blank(),text=element_text(size=18),legend.title = element_blank()) + xlab(&#39;&#39;) + 
  ylab(&#39;Number of loci&#39;) + geom_text(aes(label=n_loci_sig),position = position_dodge(width=0.9),vjust=1.1,size=8)
ggsave(p,filename = &#39;output/figures/FigureS16.png&#39;,width = 8,height = 6,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-162-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-s17---exp-of-ad-coloc-genes" class="section level2">
<h2>Figure S17 - Exp of AD coloc genes</h2>
<pre class="r"><code>#Load AD coloc genes
coloc_ad &lt;- read_tsv(&#39;output/coloc/coloc.ad.txt&#39;) %&gt;% 
  mutate(trait=&#39;Alzheimer&#39;) %&gt;% 
  filter(PP.H4.abf&gt;=0.7)</code></pre>
<pre class="r"><code>#Load Pseudo-bulk
sum_expression_ms &lt;- readRDS(&#39;data_sensitive/expression/ms_sum_expression.individual_id.rds&#39;) %&gt;% mutate(dataset=&#39;ms&#39;)
sum_expression_ad &lt;- readRDS(&#39;data_sensitive/expression/ad_sum_expression.individual_id.rds&#39;) %&gt;% mutate(dataset=&#39;ad&#39;)</code></pre>
<pre class="r"><code>#Let&#39;s merge the MS and AD dataset.
sum_expression &lt;- rbind(sum_expression_ms,sum_expression_ad)</code></pre>
<pre class="r"><code>#Get counts per million (CPM)
sum_expression &lt;- sum_expression %&gt;% 
  group_by(cell_type,individual_id) %&gt;% 
  mutate(counts_scaled_1M=counts*10^6/sum(counts)) %&gt;% 
  ungroup() %&gt;% 
  as_tibble()</code></pre>
<pre class="r"><code>#Keep AD coloc genes
sum_expression &lt;- sum_expression %&gt;% filter(ensembl%in%coloc_ad$ensembl)</code></pre>
<pre class="r"><code>m &lt;- sum_expression %&gt;% 
  group_by(cell_type,symbol) %&gt;% 
  summarise(mean_cpm=mean(counts_scaled_1M,na.rm=TRUE)) %&gt;% 
  ungroup() %&gt;% 
  spread(cell_type,mean_cpm) %&gt;% 
  column_to_rownames(&#39;symbol&#39;) %&gt;% 
  as.matrix() </code></pre>
<pre class="r"><code>pheatmap::pheatmap(log2(m+1),display_numbers = TRUE,number_color = &#39;black&#39;,angle_col = 45,cellwidth = 30,filename = &#39;output/figures/FigureS17.png&#39;,height = 5,width=7)</code></pre>
<pre class="r"><code>pheatmap::pheatmap(log2(m+1),display_numbers = TRUE,number_color = &#39;black&#39;,angle_col = 45,cellwidth = 30)</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-170-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="figure-s18---gwas-enrichment" class="section level2">
<h2>Figure S18 - GWAS enrichment</h2>
<pre class="r"><code>sum_expression_ms &lt;- readRDS(&#39;data_sensitive/expression/ms_sum_expression.individual_id.rds&#39;) %&gt;% mutate(dataset=&#39;ms&#39;)</code></pre>
<pre class="r"><code>metadata &lt;- read_tsv(&#39;data_sensitive/eqtl/covariate_pca_meta_fastqtl.txt&#39;) %&gt;% 
  column_to_rownames(&#39;id&#39;) %&gt;% 
  t() %&gt;% 
  as.data.frame() %&gt;% 
  rownames_to_column(&#39;individual_id&#39;) %&gt;% 
  as_tibble() %&gt;% 
  filter(study==&#39;MS&#39;,diagnosis==&#39;Ctrl&#39;)</code></pre>
<pre class="r"><code>#Only keep controls
sum_expression_ms &lt;- filter(sum_expression_ms,individual_id%in%metadata$individual_id)</code></pre>
<pre class="r"><code>#Keep top 1k genes that were tested using magma
entrez_2_ensembl &lt;- AnnotationDbi::toTable(org.Hs.eg.db::org.Hs.egENSEMBL) %&gt;% 
  add_count(gene_id,name = &#39;n_entrez&#39; ) %&gt;% 
  add_count(ensembl_id,name=&#39;n_ensembl&#39;) %&gt;% 
  filter(n_entrez==1,n_ensembl==1) %&gt;% 
  dplyr::select(-n_entrez,-n_ensembl,GENE=gene_id,ensembl=ensembl_id)</code></pre>
<pre class="r"><code>magma_genes &lt;- read.table(&#39;data/gwas/ms/ms.annotated_35kbup_10_down.genes.out&#39;,header=TRUE) %&gt;% 
  mutate(GENE=as.character(GENE)) %&gt;%
  inner_join(.,entrez_2_ensembl,by=&#39;GENE&#39;) %&gt;% 
  dplyr::select(GENE,ensembl)</code></pre>
<pre class="r"><code>sum_expression_ms &lt;- sum_expression_ms %&gt;% 
  group_by(cell_type,individual_id) %&gt;% 
  mutate(libsize=sum(counts)) %&gt;% 
  mutate(cpm=counts*10^6/libsize) %&gt;% 
  ungroup() %&gt;% 
  group_by(cell_type,symbol,ensembl) %&gt;% 
  summarise(mean_cpm=mean(cpm)) %&gt;% 
  group_by(symbol,ensembl) %&gt;% 
  mutate(spe=mean_cpm/sum(mean_cpm)) %&gt;% 
  ungroup() %&gt;% 
  inner_join(.,magma_genes,by=&#39;ensembl&#39;) %&gt;% 
  filter(mean_cpm&gt;1) %&gt;% 
  group_by(cell_type) %&gt;% 
  slice_max(order_by = spe,n=1000) %&gt;% 
  dplyr::select(cell_type,GENE) %&gt;% 
  mutate(cell_type=make.names(cell_type)) %&gt;% 
  write_tsv(&#39;data/magma_specific_genes/top1k.txt&#39;)</code></pre>
<pre class="bash"><code>source ~/.bashrc
cd data/magma_specific_genes

#Gene sets
top10k=&#39;top1k.txt&#39;

#GWAS
ms=&#39;../gwas/ms/ms.annotated_35kbup_10_down.genes.raw&#39;
ad=&#39;../gwas/ad/ad.annotated_35kbup_10_down.genes.raw&#39;
pd=&#39;../gwas/pd/parkinsonMeta2020.annotated_35kbup_10_down.genes.raw&#39;
scz=&#39;../gwas/scz/scz3.annotated_35kbup_10_down.genes.raw&#39;

#Run MAGMA
magma --gene-results $ms --set-annot $top10k col=2,1 --out ms
magma --gene-results $ad --set-annot $top10k col=2,1 --out ad
magma --gene-results $pd --set-annot $top10k col=2,1 --out pd
magma --gene-results $scz --set-annot $top10k col=2,1 --out scz</code></pre>
<pre class="r"><code>files &lt;- list.files(&#39;data/magma_specific_genes&#39;,pattern=&#39;.gsa.out&#39;,full.names = T)</code></pre>
<pre class="r"><code>d &lt;- tibble(files=files) %&gt;% mutate(file_content=map(files,read.table,header=TRUE)) %&gt;% 
  mutate(disease=basename(files) %&gt;% gsub(&#39;.gsa.out&#39;,&#39;&#39;,.)) %&gt;% 
  dplyr::select(-files) %&gt;% 
  unnest(file_content) %&gt;% 
  mutate(gene_set=gsub(&#39;\\.&#39;,&#39; &#39;,VARIABLE) %&gt;% gsub(&#39;OPCs   COPs&#39;,&#39;OPCs / COPs&#39;,.)) %&gt;% 
  group_by(disease) %&gt;% 
  mutate(fdr.local=p.adjust(P,method=&#39;fdr&#39;)) %&gt;% 
  mutate(sig=ifelse(fdr.local&lt;=0.05,&#39;&lt;5%FDR&#39;,&#39;Not significant&#39;)) %&gt;% 
  mutate(disease=case_when(
    disease==&#39;ad&#39; ~ &quot;Alzheimer&#39;s disease&quot;,
    disease==&#39;ms&#39; ~ &#39;Multiple sclerosis&#39;,
    disease==&#39;pd&#39; ~ &quot;Parkinson&#39;s disease&quot;,
    disease==&#39;scz&#39; ~ &#39;Schizophrenia&#39;
  ))</code></pre>
<pre class="r"><code>reorder_within &lt;- function(x, by, within, fun = mean, sep = &quot;___&quot;, ...) {
  new_x &lt;- paste(x, within, sep = sep)
  stats::reorder(new_x, by, FUN = fun)
}

scale_x_reordered &lt;- function(..., sep = &quot;___&quot;) {
  reg &lt;- paste0(sep, &quot;.+$&quot;)
  ggplot2::scale_x_discrete(labels = function(x) gsub(reg, &quot;&quot;, x), ...)
}

scale_y_reordered &lt;- function(..., sep = &quot;___&quot;) {
  reg &lt;- paste0(sep, &quot;.+$&quot;)
  ggplot2::scale_y_discrete(labels = function(x) gsub(reg, &quot;&quot;, x), ...)
}</code></pre>
<pre class="r"><code>p &lt;- ggplot(d,aes(reorder_within(gene_set,-log10(P),disease),-log10(P),fill=sig)) + 
  geom_col(position = position_dodge()) + 
  coord_flip() + 
  facet_wrap(~disease,scales = &#39;free_y&#39;) + 
  theme_classic() + xlab(&#39;&#39;) +
  scale_x_reordered() + 
  ggtitle(&#39;GWAS enrichment of top 1000 most specific genes&#39;)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-181-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggsave(p,filename=&#39;output/figures/FigureS18.png&#39;,width=8,height=5,dpi=300)</code></pre>
</div>
<div id="figure-s19---coloc-ms-gtex-and-dice" class="section level2">
<h2>Figure S19 - Coloc MS GTEx and DICE</h2>
<pre class="r"><code>coloc_heatmap &lt;- function(file,threshold=0.7,width=7,height=5,out_name){
    
  coloc_all &lt;- read_tsv(file)

  #Get best coloc if same gene in same cell type was tested in two different loci
  coloc_all &lt;- coloc_all %&gt;% group_by(ensembl,tissue) %&gt;% 
    slice_max(n=1,order_by=PP.H4.abf,with_ties=FALSE) %&gt;% 
    ungroup()
  
  #Get best coloc if same symbol in same cell type was tested multiple times (different ensembl name for same symbol at same locus)
  coloc_all &lt;- coloc_all %&gt;% group_by(symbol,tissue) %&gt;% 
    slice_max(n=1,order_by=PP.H4.abf,with_ties=FALSE) %&gt;% 
    ungroup()

  #Get absolute value of effect size of the GWAS
  coloc_all &lt;- mutate(coloc_all,beta=abs(beta_top_GWAS))
  
  coloc_all_matrix &lt;- dplyr::select(coloc_all,tissue,symbol,PP.H4.abf) %&gt;% 
  filter(!is.na(symbol)) %&gt;% 
  unique() %&gt;% 
  spread(tissue,PP.H4.abf,fill=0) %&gt;% 
  column_to_rownames(&#39;symbol&#39;)

  coloc_all_format_per_gene &lt;- coloc_all %&gt;% 
    group_by(ensembl) %&gt;% 
    filter(PP.H4.abf==max(PP.H4.abf),PP.H4.abf&gt;threshold) %&gt;% 
    mutate(closest_gene=gsub(&#39;ENSG.+ - | - ENSG.+&#39;,&#39;&#39;,closest_gene)) %&gt;% #Remove ENSEMBL name if a symbol + and ensembl gene name
                                                    #are both at same distance from GWAS top pick
    ungroup() %&gt;% 
    dplyr::rename(symbol_coloc=symbol) %&gt;% 
    filter(!is.na(symbol_coloc)) %&gt;% 
    arrange(-PP.H4.abf) %&gt;% 
    mutate(risk=case_when(
      direction ==1 ~ &#39;Up&#39;,
      direction ==-1 ~ &#39;Down&#39;,
      direction ==0 ~ &#39;Isoform&#39;
    )) %&gt;% 
    column_to_rownames(&#39;symbol_coloc&#39;) %&gt;% 
    dplyr::rename(LOEUF=oe_lof_upper_bin)
  
  coloc_all_matrix_subset &lt;- coloc_all_matrix[apply(coloc_all_matrix,1,function(x) any(x&gt;threshold)),] %&gt;% as.matrix()
  
  coloc_all_format_per_gene_direction &lt;- coloc_all_format_per_gene[rownames(coloc_all_matrix_subset),] %&gt;% 
    dplyr::select(closest_gene,risk,LOEUF,beta)
  
 ha = HeatmapAnnotation(df = coloc_all_format_per_gene_direction[,2:3,drop=FALSE],
                         which =&#39;row&#39;,
                         col = list(&#39;risk&#39; = c(&quot;Up&quot;=&quot;#E31A1C&quot;,&quot;Down&quot;=&quot;#1F78B4&quot;,&quot;Isoform&quot;=&quot;darkorange&quot;),
                                    &#39;LOEUF&#39; = circlize::colorRamp2(c(0,10),c(&#39;white&#39;,&#39;springgreen4&#39;))),
                         show_annotation_name = c(&#39;risk&#39; = TRUE,&#39;LOEUF&#39;=TRUE),
                         annotation_name_rot = 45
                         )
  
  hb = HeatmapAnnotation(df = coloc_all_format_per_gene_direction[,4,drop=FALSE],
                         which =&#39;row&#39;,
                         col = list(circlize::colorRamp2(c(min(coloc_all_format_per_gene_direction[[4]],na.rm=T),
                                                                  max(coloc_all_format_per_gene_direction[[4]],na.rm=T)),c(&#39;white&#39;,&#39;palevioletred3&#39;))) %&gt;% 
                           setNames(colnames(coloc_all_format_per_gene_direction)[4]),
                         annotation_name_rot = 45
                         )

   ht &lt;- Heatmap(coloc_all_matrix_subset,col=viridis(100),
                right_annotation=ha,column_names_rot = 45,
                left_annotation = hb,
                #row_names_side = &#39;left&#39;, 
                #row_dend_side = &#39;right&#39;,
                row_split = coloc_all_format_per_gene_direction[, 1],
                row_title_rot = 0,
                cluster_rows = FALSE,
                layer_fun = function(j, i, x, y, width, height, fill) {
                                          grid.text(sprintf(&quot;%.2f&quot;, pindex(coloc_all_matrix_subset, i, j)), 
                                                    x, y, gp = gpar(fontsize = 10,col=&quot;black&quot;))
                                                      },
                #row_title_gp = gpar(col = &#39;red&#39;, font = 2),
                #row_title = rep(&#39;&#39;,length(unique(coloc_all_format_per_gene_direction[, 1]))),
                heatmap_legend_param = list(title=&quot;PP&quot;,at = c(0,0.5,1),
                                            lables = c(0,0.5,1)))
                
  
  pdf(out_name,width = width,height=height)
  draw(ht,heatmap_legend_side = &quot;right&quot;)
  dev.off()
  
  return(ht)
}</code></pre>
<pre class="r"><code>coloc_heatmap(&#39;output/coloc/coloc.ms.gtex_dice.txt&#39;,threshold=0.7,height=30,width=10,out_name=&#39;output/figures/FigureS19.pdf&#39;)</code></pre>
</div>
<div id="figures20---genotype-pca" class="section level2">
<h2>FigureS20 - Genotype PCA</h2>
<pre class="r"><code>data &lt;- read.table(&quot;data_sensitive/genotypes/Combined_with1kg.mds&quot;,header=TRUE) %&gt;% dplyr::select(-FID) %&gt;% dplyr::select(IID,C1,C2)
ancestry &lt;- read.table(&quot;data_sensitive/genotypes/integrated_call_samples_v3.20130502.ALL.panel&quot;,header=TRUE) %&gt;% dplyr::rename(IID=sample)
d &lt;- left_join(data,ancestry,by=&quot;IID&quot;)
d &lt;- mutate(d,super_pop=ifelse(is.na(super_pop),&#39;This study&#39;,super_pop))

threshold &lt;- d %&gt;% filter(super_pop==&#39;EUR&#39;) %&gt;% summarise(meanC1=mean(C1),sdC1=sd(C1),meanC2=mean(C2),sdC2=sd(C2)) %&gt;% 
  mutate(x1=meanC1-3*sdC1,x2=meanC1+3*sdC1,y1=meanC2-3*sdC2,y2=meanC2+3*sdC2)

d &lt;- mutate(d,outlier=ifelse(C1&gt;threshold$x1 &amp; C1&lt;threshold$x2 &amp; C2&gt;threshold$y1 &amp; C2&lt;threshold$y2,&#39;keep&#39;,&#39;outlier&#39;))

p &lt;- ggplot(d) +geom_point(aes(C1,C2,col= super_pop),size=0.5) + theme_bw() + xlab(&#39;MDS1&#39;) + ylab(&#39;MDS2&#39;) + geom_rect(data=threshold,aes(xmin=x1,xmax=x2,ymin=y1,ymax=y2),alpha=0.1,color=&#39;red&#39;) + theme(text=element_text(size=18))
ggsave(p,filename=&#39;output/figures/FigureS20.png&#39;,width=7,height=5,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-185-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="figures21---eqtls-number-vs-pcs" class="section level2">
<h2>FigureS21 - eQTLs number vs PCs</h2>
<pre class="r"><code>d &lt;- read_tsv(&#39;output/eqtl/eqtl.allPCs.txt&#39;) %&gt;% 
  filter(adj_p&lt;0.05) %&gt;% 
  dplyr::count(cell_type,PCs) %&gt;% 
  mutate(PCs=gsub(&#39;PC&#39;,&#39;&#39;,PCs) %&gt;% as.numeric())</code></pre>
<pre class="r"><code>p &lt;- d %&gt;% ggplot(.,aes(PCs,n,col=cell_type)) + geom_point() + geom_line(aes(group=cell_type)) + theme_classic() + xlab(&#39;Number of expression PCs used as covariate&#39;) + guides(fill = guide_legend(reverse=T)) + ylab(&#39;Number of cis-eQTL (5%FDR)&#39;) + scale_x_continuous(breaks=seq(20, 120, 20)) + guides(col=guide_legend(title=&quot;Cell type&quot;)) + theme(text=element_text(size=18))
ggsave(p,filename=&#39;output/figures/FigureS21.png&#39;,width=8,dpi=300)
p</code></pre>
<p><img src="figure/plot_figures.Rmd/unnamed-chunk-187-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.1 (2020-06-06)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /pstore/apps/OpenBLAS/0.3.1-GCC-7.3.0-2.30/lib/libopenblasp-r0.3.1.so

locale:
[1] en_US.UTF-8

attached base packages:
 [1] grid      parallel  stats4    stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] rhdf5_2.32.4                           
 [2] viridis_0.5.1                          
 [3] viridisLite_0.3.0                      
 [4] ComplexHeatmap_2.7.6.1002              
 [5] liftOver_1.12.0                        
 [6] Homo.sapiens_1.3.1                     
 [7] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
 [8] org.Hs.eg.db_3.11.4                    
 [9] GO.db_3.11.4                           
[10] OrganismDbi_1.30.0                     
[11] GenomicFeatures_1.40.0                 
[12] AnnotationDbi_1.50.0                   
[13] Biobase_2.48.0                         
[14] rtracklayer_1.48.0                     
[15] GenomicRanges_1.40.0                   
[16] GenomeInfoDb_1.24.0                    
[17] IRanges_2.22.2                         
[18] S4Vectors_0.26.1                       
[19] BiocGenerics_0.34.0                    
[20] gwascat_2.20.1                         
[21] scales_1.1.1                           
[22] patchwork_1.0.0                        
[23] qvalue_2.20.0                          
[24] forcats_0.5.0                          
[25] stringr_1.4.0                          
[26] dplyr_1.0.0                            
[27] purrr_0.3.4                            
[28] readr_1.3.1                            
[29] tidyr_1.1.0                            
[30] tibble_3.0.1                           
[31] ggplot2_3.3.3                          
[32] tidyverse_1.3.0                        
[33] workflowr_1.6.2                        

loaded via a namespace (and not attached):
  [1] tidyselect_1.1.0            htmlwidgets_1.5.1          
  [3] RSQLite_2.2.0               BiocParallel_1.22.0        
  [5] munsell_0.5.0               withr_2.2.0                
  [7] colorspace_1.4-1            knitr_1.28                 
  [9] rstudioapi_0.11             ggsignif_0.6.0             
 [11] labeling_0.3                git2r_0.27.1               
 [13] GenomeInfoDbData_1.2.3      bit64_0.9-7                
 [15] farver_2.0.3                pheatmap_1.0.12            
 [17] rprojroot_1.3-2             vctrs_0.3.1                
 [19] generics_0.0.2              xfun_0.14                  
 [21] BiocFileCache_1.12.0        R6_2.4.1                   
 [23] ggbeeswarm_0.6.0            clue_0.3-57                
 [25] bitops_1.0-6                DelayedArray_0.14.0        
 [27] assertthat_0.2.1            promises_1.1.1             
 [29] nnet_7.3-14                 beeswarm_0.2.3             
 [31] gtable_0.3.0                Cairo_1.5-12.2             
 [33] rlang_0.4.10                GlobalOptions_0.1.2        
 [35] splines_4.0.1               rstatix_0.6.0              
 [37] acepack_1.4.1               checkmate_2.0.0            
 [39] broom_0.5.6                 BiocManager_1.30.10        
 [41] yaml_2.2.1                  reshape2_1.4.4             
 [43] abind_1.4-5                 modelr_0.1.8               
 [45] backports_1.1.7             httpuv_1.5.4               
 [47] Hmisc_4.4-0                 RBGL_1.64.0                
 [49] tools_4.0.1                 ellipsis_0.3.1             
 [51] RColorBrewer_1.1-2          Rcpp_1.0.6                 
 [53] plyr_1.8.6                  base64enc_0.1-3            
 [55] progress_1.2.2              zlibbioc_1.34.0            
 [57] RCurl_1.98-1.2              prettyunits_1.1.1          
 [59] ggpubr_0.4.0                rpart_4.1-15               
 [61] openssl_1.4.1               GetoptLong_1.0.5           
 [63] SummarizedExperiment_1.18.1 haven_2.3.1                
 [65] ggrepel_0.8.2               cluster_2.1.0              
 [67] fs_1.4.1                    magrittr_1.5               
 [69] data.table_1.12.8           openxlsx_4.1.5             
 [71] circlize_0.4.12             reprex_0.3.0               
 [73] whisker_0.4                 matrixStats_0.56.0         
 [75] hms_0.5.3                   evaluate_0.14              
 [77] XML_3.99-0.3                rio_0.5.16                 
 [79] jpeg_0.1-8.1                readxl_1.3.1               
 [81] gridExtra_2.3               shape_1.4.5                
 [83] compiler_4.0.1              biomaRt_2.44.4             
 [85] crayon_1.3.4                htmltools_0.5.1.1          
 [87] later_1.1.0.1               Formula_1.2-3              
 [89] lubridate_1.7.9             DBI_1.1.0                  
 [91] dbplyr_1.4.4                rappdirs_0.3.1             
 [93] Matrix_1.2-18               car_3.0-8                  
 [95] cli_2.0.2                   pkgconfig_2.0.3            
 [97] GenomicAlignments_1.24.0    foreign_0.8-80             
 [99] xml2_1.3.2                  vipor_0.4.5                
[101] XVector_0.28.0              rvest_0.3.5                
[103] digest_0.6.25               graph_1.66.0               
[105] Biostrings_2.56.0           rmarkdown_2.2              
[107] cellranger_1.1.0            htmlTable_1.13.3           
[109] curl_4.3                    Rsamtools_2.4.0            
[111] rjson_0.2.20                lifecycle_0.2.0            
[113] nlme_3.1-148                jsonlite_1.6.1             
[115] Rhdf5lib_1.10.0             carData_3.0-4              
[117] askpass_1.1                 fansi_0.4.1                
[119] pillar_1.4.4                lattice_0.20-41            
[121] httr_1.4.1                  survival_3.1-12            
[123] glue_1.4.1                  zip_2.0.4                  
[125] png_0.1-7                   bit_1.1-15.2               
[127] stringi_1.4.6               blob_1.2.1                 
[129] latticeExtra_0.6-29         memoise_1.1.0              </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
