---
title: "Run cell type specific eQTL analysis"
site: workflowr::wflow_site
author: "Julien Bryois"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  workflowr::wflow_html:
   toc: true
   toc_float: true
editor_options:
  chunk_output_type: console
---

```{r,echo=FALSE}
#Necessary for module load to work
knitr::opts_chunk$set(engine.opts = list(bash="--login"))
```

## Library

```{r,message=FALSE}
library(tidyverse)
library(parallel)
library(seqminer)
library(glmmTMB)
```

## Prepare data

1. Load eQTL data

```{r,message=FALSE,eval=TRUE}
d <- read_tsv('output/eqtl/eqtl.PC70.txt') %>% 
  mutate(cell_type=ifelse(cell_type=='OPCs   COPs','OPCs / COPs',cell_type))
```

```{r,eval=TRUE}
d_sig <- filter(d,adj_p<0.05)
```

2. Load gene expression

```{r,eval=TRUE}
expression_ms <- readRDS('data_sensitive/expression/ms_sum_expression.individual_id.rds') %>% mutate(dataset='ms')
expression_ad <- readRDS('data_sensitive/expression/ad_sum_expression.individual_id.rds') %>% mutate(dataset='ad')
```

```{r,eval=TRUE}
expression <- rbind(expression_ms,expression_ad)
```

Get counts per million (CPM)

```{r,eval=TRUE}
expression <- expression %>% 
  group_by(cell_type,individual_id) %>% 
  mutate(counts_scaled_1M=counts*10^6/sum(counts)) %>% 
  ungroup() %>% 
  as_tibble()
```

Compute TMM libsize

```{r,eval=TRUE}
get_tmm_libsize <- function(cell_type_id){
  
 x <-  expression %>% 
    filter(cell_type==cell_type_id) %>% 
    dplyr::select(ensembl,individual_id,counts) %>% 
    spread(individual_id,counts) %>% 
    column_to_rownames('ensembl') %>% 
    edgeR::DGEList(counts = .) %>% #TMM normalize the data using edgeR
    edgeR::calcNormFactors(.)
 
 out <- tibble(cell_type=cell_type_id,individual_id=rownames(x$samples),libsize_TMM=edgeR::effectiveLibSizes(x))
 
 return(out)
}
```

```{r,eval=TRUE}
cell_types <- unique(expression$cell_type)
tmm_libsizes <- mclapply(cell_types,get_tmm_libsize,mc.cores=8) %>% 
  bind_rows() %>% 
  mutate(individual_id=make.names(individual_id))
```

3. Load genotypes

```{r,eval=TRUE}
genotype_file <- 'data_sensitive/genotypes/processed/combined_final.vcf.gz'
snp_positions_file <- 'data_sensitive/eqtl/PC70_caveman/sig_eQTL_snps_alleles.txt'
```
  
```{r,eval=TRUE}
snp_coordinates <- read_tsv(snp_positions_file,col_names = FALSE) %>% setNames(c('chr','start','sid','A1','A2'))
```

```{r,eval=TRUE}
genotypes <- tabix.read.table(genotype_file, paste0(snp_coordinates$chr,':',snp_coordinates$start,'-',snp_coordinates$start)) %>% gather(individual_id,genotype,-CHROM,-POS,-ID,-REF,-ALT,-QUAL,-FILTER,-INFO,-FORMAT) %>% 
  dplyr::select(ID,individual_id,genotype) %>% 
    mutate(genotype_parsed=case_when(
      genotype=='1/1' ~ 2,
      genotype=='0/1' ~ 1,
      genotype=='1/0' ~ 1,
      genotype=='0/0' ~ 0,
    )) %>% as_tibble() %>% 
    dplyr::select(ID,individual_id,genotype_parsed) %>% 
    filter(!is.na(genotype_parsed)) %>% 
  dplyr::rename(sid=ID)
```

4. Get covariates

```{r,message=FALSE,eval=TRUE}
cov_files <- list.files('data_sensitive/eqtl/PC70/',pattern = 'cov.txt.gz',full.names = TRUE)

cov <- tibble(files = cov_files) %>%
  mutate(
    file_content = map(cov_files, read_tsv),
    cell_type = basename(files) %>%
      gsub('.cov.txt.gz', ' ', .) %>%
      gsub('\\.', ' ', .) %>%
      gsub('OPCs   COPs', 'OPCs / COPs', .) %>%
      gsub(' $', '', .)
  ) %>%
  dplyr::select(-files) %>%
  unnest(file_content) %>%
  gather(individual_id, cov_value, -id, -cell_type) %>%
  spread(id, cov_value) %>%
  mutate(individual_id = gsub('-|/', '.', individual_id)) %>%
  mutate_at(vars(contains('PC')), as.numeric)
```

5. Run interaction model

```{r,eval=TRUE}
#Function to try other optimizers if model did not converge
try_other_optimizers <- function(model_fit){
  
  if (model_fit$fit$convergence != 0 | model_fit$sdr$pdHess == FALSE) {
      tryCatch(
        {
          model_fit <- update(model_fit, control = glmmTMB::glmmTMBControl(
          optimizer = optim, optArgs = list( method = "BFGS" ) ))
        }, error=function(cond){
            message('Error optimizer BFGS')
        })
    }
    if (model_fit$fit$convergence != 0 | model_fit$sdr$pdHess == FALSE) {
       tryCatch(
        {
          model_fit <- update(model_fit, control = glmmTMB::glmmTMBControl(
          optimizer = optim, optArgs = list( method = "SANN" ) ))
        }, error=function(cond){
          message('Error optimizer SANN')
        })
    }
  return(model_fit)
}
```

Ambient RNA estimation.

Genes with at least 1CPM (on averafe) and for which the ambient RNA explain at most 10% of the reads will be considered expressed.

```{r,eval=TRUE}
ambient_prop_per_cell_type <- readRDS('data_sensitive/ambient/ambient.prop.estimate.droplet_utils.rds')
```

Note: 

If a gene is not expressed in a cell type, we can't estimate the genetic effect on gene expression in that cell type. We set the interaction pvalue to 0 for such cases as the genetic effect of the eQTL is by definition different between the cell type in which the eQTL was discovered and the cell type in which the gene is not expressed. 

```{r,warning=FALSE,eval=TRUE}
interaction_model <- function(i){
  
  #i <- 4220
  #i <- 631
  
  message(i)
  #write(i,'output_sbatch.txt',append = TRUE)
  
  gene_id <- d_sig$pid[i]
  ensenmbl_id <- gsub('.+_','',gene_id)
  snp_id <- d_sig$sid[i]
  cell_type_id <- d_sig$cell_type[i]
  
  expression_gene <- filter(expression,ensembl==ensenmbl_id) %>% mutate(individual_id=make.names(individual_id))
  genotype_snp <- filter(genotypes,sid==snp_id)
  
  df <- inner_join(expression_gene,genotype_snp,by='individual_id')
  
  #Add covariates
  df <- inner_join(df,cov,by=c('cell_type','individual_id'))
  
  #Add TMM libsize
  df <- inner_join(df,tmm_libsizes,by=c('cell_type','individual_id'))
  
  #Get summary stats for expression level
  gene_sumstats <- df %>% group_by(cell_type) %>% 
  summarise(mean_cpm=mean(counts_scaled_1M),
         n_samples_expressed=sum(counts>0)) %>% 
  ungroup()
  
  #Get ambient expression in the 8 cell types, only keep the ones with less than 10% of the reads that could be explained by the ambient profile.
  ambient_prop <- ambient_prop_per_cell_type %>% 
    filter(ensembl==ensenmbl_id) %>% 
    filter(mean_soup<0.1)
  
  #Get cell types expressing the gene, Same threshold as used for the eQTL analysis
  cell_type_expressed <- gene_sumstats %>% 
    filter(mean_cpm>1,n_samples_expressed>10) %>% 
    filter(cell_type%in%ambient_prop$cell_type) %>% 
    pull(cell_type)
  
  #Aggregate - Set all cell types that are not the cell type in which the eQTL was discovered as 'Others'
  df <- df %>% 
    mutate(cell_type_model=
             factor(ifelse(cell_type==cell_type_id,cell_type,'Others'),levels=c('Others',cell_type_id)))
  
  if(length(cell_type_expressed)>1){
    #Aggregate - Model formula
    full_model <- 
      as.formula(paste0('counts ~ genotype_parsed+cell_type+genotype_parsed:cell_type_model+diagnosis+study+',
                                  paste0('PC',1:3,collapse='+'),'+',
                                  paste0('PC',1:5,'_exp',collapse='+'),
                                  '+(1|individual_id)'))
  
    #Aggregate - Run model 
    full_model_nb_aggregate <- glmmTMB(full_model,offset=log(libsize_TMM),
                                     data=df,
                                     family=nbinom2,
                                     control = glmmTMBControl(parallel = 1))
  
    #if not converged, try different optimizers
    full_model_nb_aggregate <- try_other_optimizers(full_model_nb_aggregate)
  
    full_model_nb_aggregate_convergence <- ifelse(full_model_nb_aggregate$fit$convergence,FALSE,TRUE)

    nb_pvalues_aggregate <- summary(full_model_nb_aggregate)$coefficients$cond %>% 
    as.data.frame() %>% 
    rownames_to_column('variable') %>% 
    filter(grepl('genotype_parsed:',variable)) %>% 
    dplyr::select(-variable)
  } else{
    #If there is only one cell type, the pvalue is set to 0 as the genetic effect is cell type specific.
    nb_pvalues_aggregate <- tibble(Estimate=NA,`Std. Error`=NA,`z value`=NA,`Pr(>|z|)`=0)
    full_model_nb_aggregate_convergence <- TRUE
  }
  
  if(length(cell_type_expressed)>1){
    #If there is more than one cell type expressing the gene, we estimate if the genetic effect is 
    #different between the cell type in which the eQTL was discovered and all other cell types
    
    # At least one
    df <- df %>% mutate(cell_type_model=fct_relevel(cell_type,cell_type_id))
  
    ## At least one - Run
    full_model_nb <- glmmTMB(full_model,
                           offset=log(libsize_TMM),
                           data=df,
                           family=nbinom2,
                           control = glmmTMBControl(parallel = 1))

    # if not converged, try different optimizers
    full_model_nb <- try_other_optimizers(full_model_nb)

    full_model_nb_convergence <- ifelse(full_model_nb$fit$convergence,FALSE,TRUE)
  
    nb_pvalues <- summary(full_model_nb)$coefficients$cond %>% 
      as.data.frame() %>% 
      rownames_to_column('variable') %>% 
      filter(grepl('genotype_parsed:',variable)) %>% 
      dplyr::select(variable,`Pr(>|z|)`) %>% 
      mutate(variable=gsub('genotype_parsed:cell_type_model','',variable)) %>% 
      mutate(`Pr(>|z|)`=ifelse(!variable%in%cell_type_expressed,0,`Pr(>|z|)`)) %>% 
      mutate(p_bonf=p.adjust(`Pr(>|z|)`,method='bonferroni'))
  
    nb_pvalues_all <- nb_pvalues %>% dplyr::select(-p_bonf) %>% spread(variable,`Pr(>|z|)`)
    nb_pvalues_all[,cell_type_id] <- NA
    colnames(nb_pvalues_all) <- paste0(colnames(nb_pvalues_all),'_p')
  
    nb_pvalues_all <- nb_pvalues_all[,order(names(nb_pvalues_all))]
  } else {
    nb_pvalues_all <- tibble(Astrocytes=0,`Endothelial cells`=0,`Excitatory neurons`=0,
                             `Inhibitory neurons`=0,Microglia=0,Oligodendrocytes=0,
                             `OPCs / COPs`=0,Pericytes=0)
    nb_pvalues_all[,cell_type_id] <- NA
    colnames(nb_pvalues_all) <- paste0(colnames(nb_pvalues_all),'_p')
    nb_pvalues <- tibble(p_bonf=0)
    full_model_nb_convergence <- TRUE
  }
  
    out <- tibble(cell_type_id=cell_type_id,gene_id=gene_id,snp_id=snp_id,
                nb_pvalue_aggregate=nb_pvalues_aggregate$`Pr(>|z|)`,
                nb_pvalue_at_least_one=min(nb_pvalues$p_bonf)) %>% 
    cbind(nb_pvalues_all,
          nb_pvalue_aggregate_model_converged=full_model_nb_aggregate_convergence,
          nb_pvalue_at_least_one_model_converged=full_model_nb_convergence)

  return(out)
}
```

```{r,eval=TRUE}
results <- mclapply(1:nrow(d_sig),interaction_model,mc.cores=36) %>% 
  setNames(1:nrow(d_sig)) %>% 
  bind_rows() %>% 
  arrange(nb_pvalue_aggregate) %>% 
  as_tibble() %>% 
  mutate_if(is.numeric,signif,3)
```

```{r,eval=TRUE}
write_tsv(results,'output/eqtl_specific/eqtl.PC70.specific.txt')
```
